<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using OGRDBstats with partis &mdash; OGRDBstats 0.5.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using OGRDBstats with IMPre" href="impre.html" />
    <link rel="prev" title="Using OGRDBstats with IgDiscover" href="igdiscover.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> OGRDBstats
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="ungenotyped.html">Using OGRDBstats with any repertoire</a></li>
<li class="toctree-l1"><a class="reference internal" href="tigger.html">Using OGRDBstats with TIgGER</a></li>
<li class="toctree-l1"><a class="reference internal" href="igdiscover.html">Using OGRDBstats with IgDiscover</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using OGRDBstats with partis</a></li>
<li class="toctree-l1"><a class="reference internal" href="impre.html">Using OGRDBstats with IMPre</a></li>
<li class="toctree-l1"><a class="reference internal" href="all_novel.html">Producing a more complete report</a></li>
<li class="toctree-l1"><a class="reference internal" href="gapped_v.html">Creating and using a gapped V germline set</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tools/ogrdbstats_r.html">ogrdbstats.R - command-line description</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools/formats.html">Detailed Description of Input Files</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">OGRDBstats</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Using OGRDBstats with partis</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/examples/partis.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="using-ogrdbstats-with-partis">
<span id="partis"></span><h1>Using OGRDBstats with partis<a class="headerlink" href="#using-ogrdbstats-with-partis" title="Permalink to this heading"></a></h1>
<p>Although partis can provide annotated repertoires in AIRR format, some information required by OGRDBstats is only contained
in the default yaml format. A python script, <a class="reference external" href="https://github.com/airr-community/ogrdbstats/blob/master/convert_partis.py">convert_partis.py</a>, is provided for download.
This will combine output from partis’s yaml and presto annotations, producing CHANGE-O format annotations and a FASTA file of genotype V-sequences.</p>
<p>Usage of convert_partis.py:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">convert_partis</span><span class="o">.</span><span class="n">py</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="n">partis_yaml</span> <span class="n">partis_tsv</span> <span class="n">ogrdb_recs</span> <span class="n">ogrdb_vs</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">partis_yaml</span>  <span class="o">.</span><span class="n">yaml</span> <span class="n">file</span> <span class="n">created</span> <span class="n">by</span> <span class="n">partis</span>
  <span class="n">partis_tsv</span>   <span class="o">.</span><span class="n">tsv</span> <span class="n">file</span> <span class="n">created</span> <span class="n">by</span> <span class="n">partis</span> <span class="n">presto</span><span class="o">-</span><span class="n">output</span> <span class="n">mode</span>
  <span class="n">ogrdb_recs</span>   <span class="n">annotation</span> <span class="n">output</span> <span class="n">file</span> <span class="p">(</span><span class="o">.</span><span class="n">tsv</span><span class="p">)</span>
  <span class="n">ogrdb_vs</span>     <span class="n">v_gene</span> <span class="n">sequences</span> <span class="p">(</span><span class="o">.</span><span class="n">fasta</span><span class="p">)</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>   <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
</pre></div>
</div>
<p>The following shell commands will annotate with partis in both formats, merge the output, and process with OGRDBstats:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run partis to produce annotations in YAML format</span>
partis annotate --extra-annotation-columns cdr3_seqs:invalid:in_frames:stops --infname TW01A.fasta --outfname TW01A.yaml --n-procs <span class="m">5</span>
<span class="c1"># Run partis again with additional --presto-output option. This will produce TSV-formatted output from cached data</span>
partis annotate --extra-annotation-columns cdr3_seqs:invalid:in_frames:stops --infname TW01A.fasta --outfname TW01A.tsv --presto-output <span class="se">\</span>
 --aligned-germline-fname IMGT_REF_GAPPED_DEDUPED.fasta --n-procs <span class="m">5</span>
<span class="c1"># Extract and merge required information from YAML and TSV files</span>
python convert_partis.py TW01A.yaml TW01A.tsv TW01A_OGRDB.tsv TW01A_V_OGRDB.fasta
<span class="c1"># Process the resulting output to produce the genotye file and plots</span>
Rscript ogrdbstats.R --inf_file TW01A_V_OGRDB.fasta IMGT_REF_GAPPED.fasta Homosapiens TW01A_OGRDB.tsv VH
</pre></div>
</div>
<p>Although partis must be run twice - once without the presto-output option, and once with it - it will use cached information provided
other parameters remain the same, so that the overall impact on run time is low.</p>
<p>Partis by default uses germline sequences supplied
with the package. <cite>–-presto-output</cite> requires an IMGT-aligned V-gene germline file, which you may be able to obtain from the same source
that partis uses.  Otherwise, please see <a class="reference internal" href="gapped_v.html#gapped-v"><span class="std std-ref">Creating and using a gapped V germline set</span></a>. This describes a tool that can create a gapped file from the ungapped sequences.
partis will report as an error any duplicated identical sequences in the file: duplicates must be
removed before processing will complete successfully. The <a class="reference external" href="https://williamdlees.github.io/receptor_utils/_build/html/gap_sequences.html">identical_sequences</a>
command from the <a class="reference external" href="https://williamdlees.github.io/receptor_utils/_build/html/fix_macaque_gaps.html">receptor_utils</a> package will identify duplicates.</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="igdiscover.html" class="btn btn-neutral float-left" title="Using OGRDBstats with IgDiscover" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="impre.html" class="btn btn-neutral float-right" title="Using OGRDBstats with IMPre" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>