<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using OGRDBstats with TIgGER &mdash; OGRDBstats 0.5.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using OGRDBstats with IgDiscover" href="igdiscover.html" />
    <link rel="prev" title="Using OGRDBstats with any repertoire" href="ungenotyped.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> OGRDBstats
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="ungenotyped.html">Using OGRDBstats with any repertoire</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using OGRDBstats with TIgGER</a></li>
<li class="toctree-l1"><a class="reference internal" href="igdiscover.html">Using OGRDBstats with IgDiscover</a></li>
<li class="toctree-l1"><a class="reference internal" href="partis.html">Using OGRDBstats with partis</a></li>
<li class="toctree-l1"><a class="reference internal" href="impre.html">Using OGRDBstats with IMPre</a></li>
<li class="toctree-l1"><a class="reference internal" href="all_novel.html">Producing a more complete report</a></li>
<li class="toctree-l1"><a class="reference internal" href="gapped_v.html">Creating and using a gapped V germline set</a></li>
<li class="toctree-l1"><a class="reference internal" href="haplotyping.html">Haplotyping</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tools/ogrdbstats_r.html">ogrdbstats.R - command-line description</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools/formats.html">File Formats</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">OGRDBstats</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Using OGRDBstats with TIgGER</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/examples/tigger.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="using-ogrdbstats-with-tigger">
<span id="tigger"></span><h1>Using OGRDBstats with TIgGER<a class="headerlink" href="#using-ogrdbstats-with-tigger" title="Permalink to this heading">ÔÉÅ</a></h1>
<p>To conduct a V-gene analysis with TIgGER:</p>
<ul class="simple">
<li><p>Use findNovelAlleles to identify novel alleles. Write these to a FASTA file.</p></li>
<li><p>Use inferGenotype or inferGenotypeBayesian to infer the genotype.</p></li>
<li><p>Use reassignAlleles to correct allele calls in the data set, based on the inferred genotype</p></li>
</ul>
<p>The following R code, based on the <a class="reference external" href="https://tigger.readthedocs.io/en/stable/vignettes/Tigger-Vignette/">TIgGER vignette</a>, will perform these steps and save the output. It uses the sample repertoire provided with Tigger:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load packages required for this example</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tigger</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>

<span class="c1"># Save the sample germline sequences to v_germline_gapped.fasta</span>
<span class="nf">writeFasta</span><span class="p">(</span><span class="n">SampleGermlineIGHV</span><span class="p">,</span> <span class="s">&#39;v_germline_gapped.fasta&#39;</span><span class="p">)</span>

<span class="c1"># Detect novel alleles in the sample repertoire</span>
<span class="n">novel</span> <span class="o">&lt;-</span> <span class="nf">findNovelAlleles</span><span class="p">(</span><span class="n">AIRRDb</span><span class="p">,</span> <span class="n">SampleGermlineIGHV</span><span class="p">,</span> <span class="n">nproc</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>

<span class="c1"># Extract rows that contain successful novel allele calls</span>
<span class="n">novel_rows</span> <span class="o">&lt;-</span> <span class="nf">selectNovel</span><span class="p">(</span><span class="n">novel</span><span class="p">)</span>

<span class="c1"># Infer the individual&#39;s genotype, using only unmutated sequences and checking</span>
<span class="c1"># for the use of the novel alleles inferred in the earlier step.</span>
<span class="n">geno</span> <span class="o">&lt;-</span> <span class="nf">inferGenotype</span><span class="p">(</span><span class="n">AIRRDb</span><span class="p">,</span> <span class="n">germline_db</span><span class="o">=</span><span class="n">SampleGermlineIGHV</span><span class="p">,</span> <span class="n">novel</span><span class="o">=</span><span class="n">novel</span><span class="p">,</span> <span class="n">find_unmutated</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># Save the genotype sequences to a vector</span>
<span class="n">genotype_db</span> <span class="o">&lt;-</span> <span class="nf">genotypeFasta</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">SampleGermlineIGHV</span><span class="p">,</span> <span class="n">novel</span><span class="p">)</span>

<span class="c1"># Use the personlized genotype to determine corrected allele assignments</span>
<span class="c1"># Updated genotype will be placed in the v_call_genotyped column</span>
<span class="n">sample_db</span> <span class="o">&lt;-</span> <span class="nf">reassignAlleles</span><span class="p">(</span><span class="n">AIRRDb</span><span class="p">,</span> <span class="n">genotype_db</span><span class="p">)</span>

<span class="c1"># Save the repertoire with corrected allele calls to rep_genotype.tsv</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">sample_db</span><span class="p">,</span> <span class="s">&#39;rep_genotyped.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#39;\t&#39;</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>

<span class="c1"># Save the sequences used in the corrected repertoire, including novel allele sequences</span>
<span class="nf">writeFasta</span><span class="p">(</span><span class="n">genotype_db</span><span class="p">,</span> <span class="s">&#39;v_genotyped_seqs.fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>This code creates the following files:</p>
<ul class="simple">
<li><p><cite>v_germline_gapped.fasta</cite> - the germline sequences used in the annotation</p></li>
<li><p><cite>rep_genotyped.tsv</cite> - the annotated reads, with a column V_GERMLINE_GAPPED containing the corrected (genotyped) V-call</p></li>
<li><p><cite>v_genotyped_seqs.fasta</cite> - the set of germline sequences referenced in the annotated reads, including novel sequences</p></li>
</ul>
<p>These files can be provided to OGRDBstats in the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ Rscript ogrdbstats.R --inf_file v_genotyped_seqs.fasta v_germline_gapped.fasta human rep_genotyped.tsv IGHV
</pre></div>
</div>
<p>OGRDBstats will produce two files: <cite>rep_genotyped_ogrdb_plots.csv</cite> and <cite>rep_genotyped_ogrdb_report.csv</cite>.</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="ungenotyped.html" class="btn btn-neutral float-left" title="Using OGRDBstats with any repertoire" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="igdiscover.html" class="btn btn-neutral float-right" title="Using OGRDBstats with IgDiscover" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>