<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
box-sizing: border-box;
min-width: 200px;
max-width: 980px;
margin: 0 auto;
padding: 45px;
padding-top: 0px;
}
</style>


</head>

<body>

<h1 id="using-ogrdbstats">Using ogrdbstats</h1>
<p>William Lees 2023-03-08</p>
<h1 id="conducting-a-genotype-analysis-with-ogrdbstats">Conducting a
Genotype Analysis with ogrdbstats</h1>
<p><a href="https://github.com/airr-community/ogrdbstats">ogrdbstats</a>
is an R package that can be used to create an analysis of gene usage in
a adaptive immune receptor sequencing repertoire. The analysis consists
of <a href="https://github.com/airr-community/ogrdbstats/blob/master/example_ogrdbstats_genotype.csv">usage
statistics</a> and <a href="https://github.com/airr-community/ogrdbstats/blob/master/example_ogrdbstats_plots.pdf">plots</a>.
The package is intended to be used in conjunction with a tool that
infers a ‘personalised genotype’. Currently the following tools are
supported:</p>
<ul>
<li><a href="https://tigger.readthedocs.io/en/stable/">TIgGER</a></li>
<li><a href="https://igdiscover.se">IgDiscover</a></li>
<li><a href="https://github.com/psathyrella/partis">partis</a></li>
<li><a href="https://github.com/zhangwei2015/IMPre">IMPre</a></li>
</ul>
<h3 id="package-installation">Package Installation</h3>
<p>Ogrdbstats requires a recent installation of <a href="https://pandoc.org/">Pandoc</a>. If Rstudio is installed on your
machine, pandoc will already be installed. Otherwise please follow the
installation instructions on the Pandoc website.</p>
<p>Once Pandoc is installed, please install ogrdbstats from CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;ogrdbstats&#39;</span>)</span></code></pre></div>
<p>Alternatively, you can install the latest development version from
Github:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;https://github.com/airr-community/ogrdbstats&quot;</span>)</span></code></pre></div>
<p>The package requires R version 4.2.2 or above.</p>
<h3 id="using-the-package-from-the-command-line">Using The Package from
the Command Line</h3>
<p>It’s easiest to use the package from the command line. To do this,
download <a href="https://raw.githubusercontent.com/airr-community/ogrdbstats/master/ogrdbstats.R">ogrdbstats.R</a>
and copy to the directory in which you wish to conduct the analysis.</p>
<p><em>Command Syntax (see following section for detailed description of
file formats)</em></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> ogrdbstats.R [--inf_file INF_FILE] [--hap_gene HAP_GENE] REF_FILE SPECIES READ_FILE CHAIN</span></code></pre></div>
<p>Positional Arguments:</p>
<p><code>REF_FILE</code> - pathname of a FASTA file containing IMGT
gap-aligned reference germline sequences. Usually this would be <a href="https://www.imgt.org/download/GENE-DB/IMGTGENEDB-ReferenceSequences.fasta-nt-WithGaps-F+ORF+inframeP">downloaded
from IMGT</a>.</p>
<p><code>SPECIES</code> - should contain the species name used in field
3 of the IMGT REF_FILE FASTA header, with spaces removed,
e.g. Homosapiens for Human. If you are not using an IMGT REF_FILE, you
can use any single word for the species here, and the reference file
should only contain genes for that species.</p>
<p><code>READ_FILE</code> - pathname of a tab-separated file containing
the annotated reads used to infer the genotype, in MiAIRR, CHANGEO or
IgDiscover format</p>
<p><code>CHAIN</code> specifies the sequence type to be analysed. It
must be one of <code>VH, VK, VL, D, JH, JK, JL</code>.</p>
<p>Optional Arguments:</p>
<p><code>INF_FILE</code> - pathname of a FASTA file containing sequences
of inferred novel alleles. This file must be provided if the read file
contains assignments to alleles that are not listed in the reference
file.</p>
<p><code>HAP_GENE</code> - the gene to be used for haplotyping analysis
(see haplotyping section below)</p>
<p>Detailed descriptions of the required input files are given in the
next section, but for quick usage with a supported tool, please skip to
the Usage Notes for that tool towards the end of this document.</p>
<p>The package provides functions to allow you to create the reports
programmatically, or use the information it reads from input files for
your own purposes. Please refer to the ‘ogrdbstats API’ section for a
brief overview.</p>
<h3 id="detailed-description-of-input-files">Detailed Description of
Input Files</h3>
<h4 id="ref_file---fasta-file-containing-the-imgt-gap-aligned-reference-germline-sequences">REF_FILE
- FASTA file containing the IMGT gap-aligned reference germline
sequences.</h4>
<ul>
<li>All germlines that are called in the read file (apart from those of
novel alleles) should be included.</li>
<li>The sequences must be IMGT gap-aligned</li>
<li>The FASTA header can either be in IMGT’s germline library format, or
simply consist of the allele name</li>
<li>The IMGT set can be <a href="https://www.imgt.org/download/GENE-DB/IMGTGENEDB-ReferenceSequences.fasta-nt-WithGaps-F+ORF+inframeP">downloaded</a>
and used as-is: the script will filter out the records for the nominated
species. As the IMGT set changes from time to time, please make sure
that the same version is used by the inference tool and by this
script.</li>
<li>A warning will be given if any calls in the read file do not have a
corresponding sequence in the reference file or the inferred novel
alleles file. Unmutated counts will not be provided for those
sequences.</li>
</ul>
<h4 id="read_file---a-tab-separated-file-containing-the-annotated-reads-used-to-infer-the-genotype-in-miairr-changeo-or-igdiscover-format">READ_FILE
- A tab-separated file containing the annotated reads used to infer the
genotype, in MiAIRR, CHANGEO or IgDiscover format.</h4>
<ul>
<li><p>The format will be determined automatically by the
script.</p></li>
<li><p>MiAIRR format files must contain at least the following columns:
<code>sequence_id, v_call_genotyped, d_call, j_call, sequence_alignment, cdr3</code>.
For J or D inferences they must also contain
<code>J_sequence_start</code>, <code>J_sequence_end</code>,
<code>J_germline_start</code>, <code>J_germline_end</code>, or the
equivalent fields for D genes. <a href="https://www.ncbi.nlm.nih.gov/igblast/">IgBLAST</a>’s `--format
airr` creates compatible MiAIRR format files.</p></li>
<li><p>CHANGEO files must contain at least the following columns:
<code>SEQUENCE_ID, V_CALL_GENOTYPED, D_CALL, J_CALL, SEQUENCE_IMGT, CDR3_IMGT</code>,
<code>V_MUT_NC</code>, <code>D_MUT_NC</code>, <code>J_MUT_NC</code>,
<code>SEQUENCE</code>, <code>JUNCTION_START</code>, <code>V_SEQ</code>,
<code>D_SEQ</code>, <code>J_SEQ</code>. If you would like to process
files from IMGT V-Quest, please <a href="https://changeo.readthedocs.io/en/stable/examples/imgt.html">parse
them with CHANGEO</a> to convert them to CHANGEO format.</p></li>
<li><p>D- related fields are only required for heavy chain
records.</p></li>
<li><p>In both the above file formats,
<code>v_call_genotyped/V_CALL_GENOTYPED</code> should contain the V
calls made after the subject’s V-gene genotype has been inferred
(including calls of the novel alleles). Sequences may be either unagpped
or IMGT gap-aligned. Determining the personalised V-gene genotype is
recommended when processing D or J gene inferences, so that V-gene usage
counts are accurate. However, this step can be omitted for D or J gene
processing by providing a V_CALL field instead of
V_CALL_GENOTYPED.</p></li>
<li><p>For IgDiscover, the file ‘final/filtered.tab’ should be used -
see section on IgDiscover below.</p></li>
</ul>
<h4 id="inf_file---fasta-file-containing-the-inferred-novel-alleles">INF_FILE
- FASTA file containing the inferred novel alleles</h4>
<ul>
<li>Sequences in the inferred file should all be of the same type: VH,
VK, VL, D, JH, JK, or JL</li>
<li>The header should simply consist of the allele name as assigned by
the tool.</li>
<li>V-gene sequences may either be IMGT gap-aligned or not aligned. If
they are not aligned, the script will determine the nearest reference
gene and use it as a template. If you are not satisfied with the
resulting alignment, just align the sequence in the inferred file as you
prefer.</li>
<li>If a gene with the same name is present in both the germline file
and the inferred file, its presence in the inferred file will be
ignored. This makes it easier to use the script with inference tools
that do not write the inferred sequences to a separate file.</li>
</ul>
<h3 id="output">Output</h3>
<ul>
<li><code>&lt;READ_FILE&gt;_ogrdb_report.csv</code> - the Genotype File
ready to be uploaded to OGRDB.</li>
<li><code>&lt;READ_FILE&gt;_ogrdb_plots.csv</code> - plots (see next
section for details).</li>
</ul>
<p><code>READ_FILE</code> is used as a prefix to the output file names.
They will be written to the directory containing the read file.</p>
<p>If you are submitting inferences to OGRDB, you will be prompted to
upload the genotype file. Please also upload the plots file as an
attachment to the Notes section of your submission.</p>
<h3 id="plots">Plots</h3>
<p>The script produces the following plots:</p>
<ul>
<li>For each allele used in the the read file, a histogram showing the
number of mutated and unmutated sequences</li>
<li>Barcharts showing nucleotide usage at locations in the IMGT-aligned
sequence: both across the sequence as a whole, and in more detail at the
3’ end</li>
<li>A barchart showing usage of the alleles of potential haplotyping
genes, across the whole genotype. This can be used to identify a
suitable gene for haplotyping analysis.</li>
<li>For each potential haplotyping candidate (selected by the tool from
the usage chart above), a plot comparing the usage of the two most
frequently used alleles of that gene.</li>
</ul>
<p>The nucleotide usage plots are not produced from IgDiscover output,
as aligned V-sequences are not available.</p>
<h3 id="haplotyping">Haplotyping</h3>
<p>The script should first be run without the optional
<code>HAP_GENE</code> parameter. If, having consulted the plots, you
identify a suitable gene for haplotyping, please run the script again,
with this gene specified as <code>HAP_GENE</code>. The haplotyping_gene
and haplotyping_ratio columns of the genotype file will be appropriately
populated. A J-gene should be used with V- and D- gene inferences, and a
V-gene with J-gene inferences.</p>
<h3 id="example">Example</h3>
<p>To download the IMGT reference file and complete an analysis using <a href="https://github.com/airr-community/ogrdbstats/tree/master/testdata/VH_tigger">example
files</a>, run the following commands:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> IMGT_REF_GAPPED.fasta https://www.imgt.org/download/GENE-DB/IMGTGENEDB-ReferenceSequences.fasta-nt-WithGaps-F+ORF+inframeP</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> ogrdbstats.R <span class="at">--inf_file</span> TWO01A_naive_novel_ungapped.fasta <span class="at">--hap_gene</span> IGHJ6 IMGT_REF_GAPPED.fasta Homosapiens TWO01A_naive_genotyped.tsv VH</span></code></pre></div>
<h3 id="usage-notes">Usage Notes</h3>
<p>Usage notes are indicative only and are not intended to discount
other approaches. Notes for other tools will follow.</p>
<h3 id="usage-notes---tigger">Usage Notes - TIgGER</h3>
<p>To conduct a V-gene analysis with TIgGER:</p>
<ul>
<li>Use <code>findNovelAlleles</code> to identify novel alleles in a
Change-O-formatted data set. Write these to a FASTA file.</li>
<li>Use <code>inferGenotype</code> or <code>inferGenotypeBayesian</code>
to infer the genotype.</li>
<li>Use <code>reassignAlleles</code> to correct allele calls in the data
set, based on the inferred genotype</li>
<li>Provide the resulting Change-O file, together with the FASTA file
containing the novel alleles, to <code>ogrdbstats</code>.</li>
</ul>
<p>Note that <code>inferGenotype</code> will not necessarily include
every inferred allele produced by <code>findNovelAlleles</code> in the
genotype that it produces. Only those alleles included in the genotype
will be considered by <code>genotype_statistics.R</code> because,
leaving other considerations aside, no sequences are assigned to other
alleles.</p>
<p>TIgGER provides additonal information, including its own plots and
statistics We encourage you to take these into consideration, and to
upload them as attachments to your submission if they are
informative.</p>
<h3 id="usage-notes---igdiscover">Usage Notes - IgDiscover</h3>
<p>Following an IgDiscover run, please copy ogrdbstats.R to IgDiscover’s
<code>final</code> directory. The commands below can then be used to
download the IMGT reference file and run a VH gene analysis. All
commands should be run in the <code>final</code> directory.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> wget <span class="at">-O</span> IMGT_REF_GAPPED.fasta https://www.imgt.org/download/GENE-DB/IMGTGENEDB-ReferenceSequences.fasta-nt-WithGaps-F+ORF+inframeP</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> unzip final.tab.gz</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> Rscript ogrdbstats.R <span class="at">--inf_file</span> database/V.fasta IMGT_REF_GAPPED.fasta Homosapiens filtered.tab VH</span></code></pre></div>
<p>alternatively, to produce a JH gene analysis:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> Rscript ogrdbstats.R <span class="at">--inf_file</span> database/J.fasta IMGT_REF_GAPPED.fasta Homosapiens filtered.tab JH</span></code></pre></div>
<h3 id="usage-notes---partis">Usage Notes - partis</h3>
<p>The information required by generate_statstics.R is split between
partis’s normal yaml output and that provided by the ‘presto-output’
mode. A python script, <a href="https://github.com/airr-community/ogrdbstats/blob/master/convert_partis.py">convert_partis.py</a>,
is provided. This will combine output from partis’s yaml and presto
annotations, producing CHANGEO format annotations and a FASTA file of
genotype V-sequences. These files can then passed to
generate_statistics.R. convert_partis.py is written in python 2.7 for
compatibility with partis, and can be run from the command line in the
same virtual environment.</p>
<p>Usage of convert_partis.py:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> convert_partis.py [-h] partis_yaml partis_tsv ogrdb_recs ogrdb_vs</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">positional</span> arguments:</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">partis_yaml</span>  .yaml file created by partis</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">partis_tsv</span>   .tsv file created by partis presto-output mode</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ogrdb_recs</span>   annotation output file <span class="er">(</span><span class="ex">.tsv</span><span class="kw">)</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ogrdb_vs</span>     v_gene sequences <span class="er">(</span><span class="ex">.fasta</span><span class="kw">)</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="ex">optional</span> arguments:</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">-h,</span> <span class="at">--help</span>   show this help message and exit</span></code></pre></div>
<p>Although partis must be run twice - once without the presto-output
option, and once with it - it will use cached information provided other
parameters remain the same, so that the overall impact on run time is
low. Typical processing steps are shown below. Note that –presto-output
requires an IMGT-gapped V-gene germline file. This can be extracted from
the full germline library downloaded from IMGT (see ‘Prerequisites’
above), but partis will report as an error any duplicated identical
sequences in the library: duplicates must be removed from the file
before processing will complete successfully. Note in the examples below
the –extra-annotations option. The CDR3 is required by
generate_statistics.R: the other fields are included for reference.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run partis to produce annotations in YAML format</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">partis</span> annotate <span class="at">--extra-annotation-columns</span> cdr3_seqs:invalid:in_frames:stops <span class="at">--infname</span> TW01A.fasta <span class="at">--outfname</span> TW01A.yaml <span class="at">--n-procs</span> 5</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Run partis again with additional --presto-output option. This will produce TSV-formatted output from cached data</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="ex">partis</span> annotate <span class="at">--extra-annotation-columns</span> cdr3_seqs:invalid:in_frames:stops <span class="at">--infname</span> TW01A.fasta <span class="at">--outfname</span> TW01A.tsv <span class="at">--presto-output</span> <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a> <span class="at">--aligned-germline-fname</span> IMGT_REF_GAPPED_DEDUPED.fasta <span class="at">--n-procs</span> 5</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract and merge required information from YAML and TSV files</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> convert_partis.py TW01A.yaml TW01A.tsv TW01A_OGRDB.tsv TW01A_V_OGRDB.fasta</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Process the resulting output to produce the genotye file and plots</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> ogrdbstats.R <span class="at">--inf_file</span> TW01A_V_OGRDB.fasta IMGT_REF_GAPPED.fasta Homosapiens TW01A_OGRDB.tsv VH</span></code></pre></div>
<h3 id="usage-notes---impre">Usage Notes - IMPre</h3>
<p>IMPre does not provide a set of sequences annotated with the novel
allele calls. The sequences must be annotated by a separate tool in
order to provide the information needed for the OGRDB genotype. One
possible approach is as follows:</p>
<ul>
<li>Annotate with IgBLAST using a custom germline set that includes the
novel alleles inferred by IMPre. Details for creating IgBLAST’s germline
database are given in the <a href="https://ncbi.github.io/igblast/cook/How-to-set-up.html">setup
notes</a>. The novel alleles inferred by IMPre can be added to the
germline sequences downloaded from IMGT, before running
makeblastdb.</li>
<li>Select the AIRR output format, and provide the resulting annotion
file to genotype_statistics.R, along with the novel inferences provided
by IMPre</li>
</ul>
<h3 id="ogrdbstats-api">ogrdbstats API</h3>
<h4 id="creating-reports-programmatically">Creating reports
programmatically</h4>
<p><code>generate_ogrdb_report()</code> takes equivalent arguments to
ogrdbstats.R and generates the genotype file and plots.</p>
<h4 id="customising-and-extending-the-reports">Customising and Extending
the Reports</h4>
<p><code>read_input_files()</code> reads and parses the input files. It
returns a representation of the genotype file in memory, as well as a
number of other structures containing related information.</p>
<p><code>make_barplot_grobs()</code>,
<code>make_novel_base_grobs()</code> and <code>make_haplo_grobs()</code>
create the plots: each function returns a list of grobs that can be used
as you wish, or passed to <code>write_plot_file()</code> to create a the
standard file of plots.</p>
<p>The use of the grob functions is demonstrated below with example data
included in the package.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ogrdbstats)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>reference_set <span class="ot">=</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/ref_gapped.fasta&quot;</span>, <span class="at">package =</span> <span class="st">&quot;ogrdbstats&quot;</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>inferred_set <span class="ot">=</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/novel_gapped.fasta&quot;</span>, <span class="at">package =</span> <span class="st">&quot;ogrdbstats&quot;</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>repertoire <span class="ot">=</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/ogrdbstats_example_repertoire.tsv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;ogrdbstats&quot;</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>rd <span class="ot">=</span> <span class="fu">suppressMessages</span>(</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_input_files</span>(reference_set, inferred_set, <span class="st">&#39;Homosapiens&#39;</span>, repertoire, <span class="st">&#39;IGHV&#39;</span>, <span class="cn">NA</span>, <span class="st">&#39;V&#39;</span>, <span class="st">&#39;H&#39;</span>, <span class="cn">FALSE</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>barplot_grobs <span class="ot">=</span> <span class="fu">make_barplot_grobs</span>(rd<span class="sc">$</span>input_sequences, rd<span class="sc">$</span>genotype_db, rd<span class="sc">$</span>inferred_seqs, </span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>                                   rd<span class="sc">$</span>genotype, <span class="st">&#39;V&#39;</span>, rd<span class="sc">$</span>calculated_NC)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>base_grobs <span class="ot">=</span> <span class="fu">make_novel_base_grobs</span>(rd<span class="sc">$</span>inferred_seqs, rd<span class="sc">$</span>input_sequences, <span class="st">&#39;V&#39;</span>, <span class="cn">FALSE</span>)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs=</span><span class="fu">list</span>(barplot_grobs[<span class="dv">3</span>][[<span class="dv">1</span>]], base_grobs<span class="sc">$</span>end[<span class="dv">1</span>][[<span class="dv">1</span>]], </span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>                                   base_grobs<span class="sc">$</span>conc[<span class="dv">1</span>][[<span class="dv">1</span>]]),<span class="at">ncol=</span><span class="dv">1</span>)</span></code></pre></div>
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAeAAAAWgCAMAAACxHWDYAAABL1BMVEUAAAAAADoAAGYAOjoAOmYAOpAAZrYbnnczMzM6AAA6ADo6AGY6Ojo6OmY6OpA6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOgBmOjpmZmZmkJBmkLZmkNtmtrZmtttmtv9uTU1uTW5uTY5ubk1ubqtujo5uq+R1cLOOTU2OTW6OTY6Ojo6OyP+QOgCQOjqQZjqQZpCQkGaQkNuQttuQ29uQ2/+rbk2rbm6r5P+2ZgC2Zjq2kDq2kGa2kJC2tma2ttu225C229u22/+2///Ijk3I5KvI///ZXwLbkDrbtmbbtpDb25Db27bb29vb2//b/7bb///kq27kyMjk///nKYr/tmb/tpD/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///8sbnVZAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO3dCXsjx3Wv8aak4YixfK0QoxnqKokcz2ixNOOQdnyToQhnlcSMKDlRCNMIqJAE+/t/htvVG6qbvZzqDd0H7/8ZiQu6Dqrqx16xeT5RHW/bHSD9BmDlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHkAVp5mwP/5uwPPe/vjS3GD26P9hy0v9l6Hv7s+2L89euvbdNH426X3fFNg/TvP29vcYXzb7VH4Zf3PB9aN8S83WXppnmcqWe0yJX74q2DRn38jHt2Y0wR4/Yd4umIgQWLgbEsL2J97h/GiF/F31wcW8O1HYbP4ryS+bf3j7Yev1z/6QWPrxmDRKmC70rzw2/VJvPChryBNgOfe278K/tb/9w9eutrVJQbOtrSBrw8eRavP+iT6rZnmDdSF9/6l/9NHcYP4tuuDt997L0BZeu98E9wYLf7DgZcHjprE9a1KVju7xNx75+ugwfcH8r/fEacB8PXBO7HrRdFUFiYCzrW0gRPX8Idwib3fb6rHPMn9JbcFfyiP/u3Sn4dNw4br33rv/L4K2K5ktbO+XXrxn8Iy3WBMOQ2AN6y3f/+Pwf9/SnZqweT98SNv75e+/ct4yRA419IGTrfM82iZYCe7zEP9Md47p7fdHv3Nwa8Ss/DL7YcfXz5o5/vWGrypZLWzS8yTFXf9L1+7T87o4g6cnapwfxjvvpKd16H9yyghcL5lBvj2KLwx/bLv56AuPO/RN37mtv953794/zI5gJsXHZwV9TquZLWzvs13cupxB04PiOPMvb+8NEdPzw3w/mUwffv2L61G+ZYX6bGP+X205iyjFdlg5aDm73nRxjN/mzNwXKkYON/Jqacx8O2RkXl0Ge8z1ydmP2qm2KyD1i+tRrmWOeBoHxqWiLblD6HCv50HtzkDx5UALk48VRvgGCnZk5n/W7+MGtmb6A1w5rgqtI2+vT4wq/EykrTOV8wi1m2+Vdt3AzaV2ESXJD0KyVomc5P7ZbRkNIPZljngcNWMlkhX7b3XGWADYN1m17IIJcAWZelBlvXdhNPoNCmZq8gynX4bOHeRIDlNslvmgYNl4kOsB4jpoVcRsH2OY1IJbFUqOU1KO7np7ZTT7ELH3sc/BnP2w0fhdnnvl8E8fB+eyybA1i+jpBc6rJZ54GCt+lfbxoaah5cnTtJ1eZld0LrQUbcGW5XKL3T8Q3TtUnqWP+Y0uhadXHA0s5KcEQVYG2Drl1GSnVymZR44aGRfGsvuZzO79Bxi5lJlDbBdiUuVZfnJPGSw94uv4x+CyfjFN/Ym2vpllPTY1G6ZB16fZOZU8GBDcuM/2Q821OyD7QcbNu3sEuvvg/Mo7+caLnPwcKH6AKw8WoGthwg1nOw0D8DKoxWYxAFYeQBWHoCVpxJ4NZs9OfP9+9PZB2+G6hDpNlXAN5+c+auA9vyZ+UcmmbpN9N2rs+Cff/Mpq/A0Uwe8eHplcA1ykHeDRC2+E8Vx0ceyOA7RoasKUw188+JJtPbGwJsWAE8kdWtwsA8GeMoR7IMBnnLcD7IAnlSqgM0p0s3nV7nTJIAnlco1eDELr3DcvbQvdAA8qbhfqgR4UgEY4MIWAE8kAANc2ALgiQRggAtbfClKuKh8gmVLulo4dFVhAAa4sAXAEwnAABe2AHgiARjgwhYATyQAA1zYAuCJBGCAC1sAPJEADHBhC4AnklEBy4qGVR0edwK4SQuAJxKAAS5sAfBEAjDAhS0AnkgABriwBcATCcAAF7YAeCIBGODCFgBPJAADXNhCJbCsqOtLZ7YbgAEubAHwRAIwwIUtAJ5IAAa4sAXAEwnAABe2AHgiARjgwhYATyQAA1zYYteBJ/O3ADDAhS0ABhhggAHuPQADXNgCYIABBhjg3gMwwIUtAAYYYIAB7j0ANwOWDeq79kBtAzDAhS0ABhhggAHuPQADXNgCYIABBhjg3gMwwIUtAAYYYIAB7j0AA1zYAmCAAQYY4N4DMMCFLQAGGGCAAe49AANc2AJggAEGGODeAzDAhS0AFo/Koas9BGCAC1sADDDAAAMMMMDtAjDAhS0ABhhggAEGGOB2ARjgwha7DtxPV3sIwM5FAc6OD2CAAQa4xax1XRTg7PgABhhggFvMWtdFAc6OD2CAAQa4xax1XRTg7PgABhhggFvMWtdFAc6OD2CAAQa4xax1XRTg7PgABhhggFvMWtdFAc6OD2CAAd4S8M2L2ezY9+9PZx+8ybWQjw/g0QLfvTz2V0/O/PNn5l+2hXx8AI8WePX0Klh7j+9enfk3n77JtpCPD+DRApsEa7HBNciZFvLxATxq4MXTKwv43SAA99fVHlIDvAiOrliD9QIvgkMsgPUCL8KzIw6ytALffBKttpwmKQU+n5kcB0fSXOhQCVzVQj4+gAEGGOAWs9Z1UYCz4wMYYIABbjFrXRcFODs+gAEGGOAWs9Z1UYCz4wMYYIABbjFrXRcFODs+gAEGGOAWs9Z1UYCz4wMYYIABbjFrXRcFODs+gAEGGOAWs9Z1UYCz4wMYYIABbjFrXRcFODs+gLsAbtQI4EZFAc6OD2B5VwFuMmtdFwU4Oz6A5V0FuMmsdV0U4Oz4AJZ3FeAms9Z10ZECL/deA9xJ0XECr0/+eh/gToqOE/j6Z//+s28B7qLoOIEv9v35IcBdFB0l8O2Hr/3lo8tqYLOQyfKtonUdYLtqP11tDmxwb4+ey4AvAK6t2k9XGwOvTzyTgsOsDfDcS1O0LQc4U7WfrjYGvg4PsK7/4uGZUtEaXB2ARwg8j9bdgsMsDrKci44QuOLoKQMcb8k99sG1VfvpalPgWq4480pZgK2q/XS1Z+Dbo8qjK4Ctqv10tXfggvMogAur9tPVnoHXJ4WXqwEuqNpPV3tIZh+89ASrMMD9dbWHZDfRHkfRIwUunzQHYFkABjgzPoBHA8wmWjlwnOpHCwGePHBy5RpgrcA84K8cmAf8Jwh8fVD4eH8R8O0Rm+jJAZutbuGVyKKj6KLnbgE8auDoYYS6Z3TIAvAIgcsPmwB2LjpK4NKtbhbY7Km94te4AKwAeBk+n/Ki+jElgMcILNpEr0+iZ3RcVB5lATxC4Oggq+gZG0XP6OBCx+SA/Ytgx7o+KVgzWYOdi44SONi9Ci50sA+eLnA1VxKOopUDSwIwwJnxATwW4Oha9fVBzbPfAZ4o8O1RdPCcfAV4TMCNY70+OHGteQULwP11tTzl5aXAyUmwz3mwTmDrMhdXsnQCb9ZggBUCW8+lnLOJ1gh8ffA89w3AqoDjK9H+RfWb7AA8WeD4LTrqXiMM8CiBw2dMFqyaXKp0LjpK4HC/WnQFA2DnomMEjq9iXD98w1mAnYuOEbhAFuCmRccILH3aLMBTBp4XPVkDYOeiYwSON9EF7zYKsHPRMQLHB1kAd1F0jMDR8zTWJ2yiOyg6SuDoKlXNy0cBnjBwNRfAAAMMcEkHAAYYYICbz1rXRaf6tFm3FvLxAdwF8J8LAzDAADsXBTg7PoABBhjgFrPWdVGAs+MDGGCAAW4xa10XHSFw+N4qxe+uArBz0REC++XPqwTYuSjA2fEBDDDAALeYta6LApwdH8AA6waWjx9gVwuAAXYvCnB2fAAPAlzNBTDAAANcPD6AAQZY9M7AjQKwc1GAs+MDeMzAN5++8f3709kHb3It5OMDeMTAdy+N7Pkz8y/bQj4+gMcLvJj9OliD716dRWuy3UI+PoDHC/ynKwN7EyMHeTcIwI5dHTFwtA+2gDct5OMDGGCAAW5m4djVHQXmIEs5MKdJ2oHvXnKhQytwaQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+AAGGGCAm1k4dhXgTQv5+HYeWL5kHwHYuSjAuV4DLF2yjwDsXBTgXK8Bli7ZRwB2LgpwrtcAS5fsIwA7FwU412uApUv2EYCdiwKc6zXA0iX7CMDORQHO9Rpg6ZJ9BGDnogDneg2wdMk+ArBzUYBzvQZYumQfAdi5KMC5XgMsXbKPAOxcFOBcrwGWLtlHAHYuCnCu1wBLl+wjADsXBTjXa4ClS/aRqQLLizp01eH+O1+0LWQ1l3sLh14DLF2yjwBsFXW4/84XbQtZzeXewqHXAEuX7CMAW0Ud7r/zRdtCVnO5t3DoNcDSJfsIwFZRh/vvfNG2kNVc7i0ceg2wdMk+ArBV1OH+O1+0LWQ1l3sLh14DLF2yjwBsFXW4/84XbQtZzeXewqHXAEuX7CMAW0Ud7r/zRdtCVnO5t3DodS/ADvcvX/TPsjh2VV60jwBsLQrwpoXDBAMMMMAA56o63L98UYA3LRwmGGCAAQY4V9Xh/uWLArxp4TDBAAMMMMC5qg73L18U4E0LhwkGGGCApwwsX9Rh1hzuXz7BAG9aOEywfFGHWXO4f4CbtHCYYPmiDrPmcP8AN2nhMMHyRR1mzeH+AW7SwmGC5Ys6zJrD/QPcpIXDBMsXdZg1h/sHuEkLhwmWL+owaw73D3CTFg4TLF/UYdYc7h/gJi0cJli+qMOsOdw/wHW5P5198CbXwmGC5Ys6zJrD/QNcl/Nn5l+2hcMEyxd1mDWH+we4JnevzvybT9NVGGDHro4e2OAa5EwLhwmWL+owaw73D3BNLOB3g3hlKb9lmEVV3P+WgSvzrvxee1lU5/23DcAjv/+2cT/IKo3OCd72/beN+2kSmVQkwHcv7QsdZFLp/rCNjCoAKw/AyjNS4NVs9iQ8MQsP329ezGZPr7bdp9aJB2VGc5x/CKe3jBP45pMzf2WGf/fS/H+l4hg+HtTdy2N/FUAPdG4yTmATc/q9mP3arMHnX2y7M10lGNQq2Bbdnx5Lry60zXiBF8FE/OnKzMH9V/WX0SaSRbSnCdZi6fXBthkr8M2LzT747tVn8R554kkGZZx3HdiP9sEhsNl53f1GxaWWaFCL4P8AR6NPdlNDzMUACYexCI+lAdYKvAjX4t0+yDJbspvPzfGImYPNT5NOPAyzwzHZ7dOkxSy+ChD+kac/TTvRMM5nJsdDPYQzUmDSVQBWHoCVB2Dl6Qo4OGSYzWSXjM2jKcLDR4eqK/G1Loei8qoORV3uv306Ar55cWx6Lum2OVu4PxUJO1S9P/074R+NQ1F5VYeiLvffQboBvj8N+ys5cTePlkUPnXVZNVjoP2RXDVyKiqs6FHW6/w7SDbBDd1fykz+XSVg8Ez6o6DSz0qoORQeTjdMN8Er+fIt+FjUX/mSLOxSVV+1nUJ2kQ+DzmeRBPde5kFeNNv4dFpVXdR2/8P67SKebaMkjAu6baEHV+9Pw+p/kgMihq/KqDkUdFu0knR5kSXodrRGi9UJeNZo1lyM3uYWkqkNRh0U7SWenSV+Yvku2O+bR0PtT0YZaXDV+WEZ2QCTuqkNVh/E7LNpFurrQEW7OZOeMK/GS0qrJ6iDb/Eu76lTVYfwOi3YQLlUqD8DKA7DyAKw8ACsPwMoDsPIArDwqgOfRm4jtF9x0e/Tc95fPo6+7GB3Ajy59Yxl+eZidxTVRBOwv914X3gzw1BMDXx+YzXGwrT4Mfwi+CTfN5rtH/x0ixzfeHv3tkblxF6IJ2KzByxB1P8F+btbe5L/0xtujYMmLkvVdWRQBG7r1iVl7A+nlW9+GN9nA6Y23R4fx+q4/OoCjo+hULfj/7VEkbANbNz7fmT2zDuD08DkyjFbXZB+cAw6+A3hiyQMnW9+52RqzBk8/G+B0Nxv+FK+sBftggCeVDXB6oBwaL5M1+DB/FA3wpGIBp+fB5muAHDLOH5wHA0yUBGDlAVh5AFYegJUHYOUBWHkAVh6AlQdg5QFYeQBWHoCVB2DlAVh5AFYegJUHYOUBWHk6A/7P3x143tsfF7++ryjm2W8PWiavJ7k+2E+euh4uGn+7tF9QtP6d5+19nHm6nZ8+1Wr9zwfWjQ+ef7X00jzPVLLaZUr88FfBoj//Rjy60aQj4PUf4umSv+AnBs62tID9efjkOZOL+LvrAwv49qPMi4Kj29Y/3n74ev1j8mKH/XTRKmC70rzw2/BZ9PGLJyaWjoDn3tu/Cv7W//cPXrra1SUGzra0ga8P4idLrk+i35pp3kBdeO9f+j99FDeIb7s+ePu99wKUpffON8GN0eI/HBS+lHB9Ete3Klnt7BJz752vgwbfH0zvBWvdAF8fvBO7XohflRkB51rawIlr+EO4xN7vN9VjnuT+ktuCP5RH/3ZpXtOQNFz/1nvn91XAdiWrnfXt0ov/FJaF7yIw6nQDvGG9/ft/DP7/U7JTCybvjx95e7/07V/GS4bAuZY2cLplnkfLBDvZZR7qj/HeOb3t9uhvDn6VmIVfbj/8+PJBO9+31uBNJaudXWKerLjrf/m66RRtK50AZ6cqfvF1uPtKdl6H9i+jhMD5lhng+D0Z0i/7fg7qwvMefeNnbvuf9/2L9y+TA7h50cFZUa/jSlY769t8JyeVToDTA+I4c+8vL83RU/gav/3LYPr27V9ajfItL9JjH/P7aM1ZRiuywcpBzd/zoo1n/jZn4LhSMXC+k5NKl8C3R0bm0WW8z1yfmP2omWKzDlq/tBrlWuaAo31oWCLalj+ECv92HtzmDBxXArg48VRtgGOkZE9m/m/9Mmpkb6I3wJnjqtA2+vb6IHrBUfwuHOn5ilnEus23avtuwKYSm+iSpEchWctkbnK/jJaMZjDbMgccrprREumqvfc6A2wArNvsWhahBNiiLD3Isr6bSro6TUrmKrJMp98Gzl0kSE6T7JZ54GCZ+BDrAWJ66FUEbJ/jmFQCW5VKTpPSTm56O5l0dqFj7+Mfgzn74aNwu7z3y2Aevg/PZRNg65dR0gsdVss8cLBW/attY0PNw8sTJ+m6vMwuaF3oqFuDrUrlFzr+Ibp2ObmXnHZ1LTq54GhmJTkjCrA2wNYvoyQ7uUzLPHDQyL40lt3PZnbpOcTMpcoaYLsSlyrL8pN5yGDvF1/HPwST8Ytv7E209cso6bGp3TIPvD7JzKngwYbkxn+yH2yo2QfbDzZs2tkl1t8H51Hezyd3mYOHC9UHYOXZIWDrIcLJnew0D8DKs0PAuxmAlQdg5QFYeWqAw4/QvT+dfbD5spjNjof/KHolWUUf/R1P5RCpBr57afpx/iz8KN3oy92rM/Mp5eeSj1MnuZgPmzafVhvP6BCpBF7Mfh2sqOZzdIMVNv4S//tc8gHPpCDBPMZTOcjdVQL/6cp04yZCjr9Ea/BmBX43SFjku1w8L/tzwSL5mDKPqyNYxCzzZU2E9yTosKAz0SQlE7Z4ehVPZZeOpRHsg7PA4T44twIXAhtiLzsZOwu8meWbF0/O/HED++YY4XgVHmrZRQpnwCIuW2S3gMNPjB898OpZ/LNdpGQO0tW4fJHdAs5OZe8RAGcPsoJf3n91Fv9sFymfhsi4cpHdAh7NQVZ8ups5TfLNCuzL1+DEGGA/3Dz75vhlLKdJMfDdy/C0PP4SbVyE+2Cb2APYHKHaUzlAOrlUKQEOJsPLnzntHvDwGRLYf3hyDHDfGRg4f3IMcN8ZHLjK2JQBuNtsA/i7sk21KQNwt9kScPFqbMoA3G26AxbMaW56HhibMgB3m20CPzAWlDGLAOyQLQN/l388AuCOs33g7OMRAHecMQB/Zz0eAXDHGQlwZCwoYxYB2CHjAQ6NAe46owIOH48AuNOMDDgoU21sFgHYIeMDDlJhbBYZDlhQBuAGwBXGZhGAHTJW4NAY4PYZMXDhamwWAdghowb+8uFqbBYB2CFjBw6IAW6T8QNnhc0iADtkAsAZYbMIwA6ZArAtbBYB2CGTALaEzSIAO2QawE4yANuZCLBnVQHYJRMBToXNIgA7ZCrAibBZBGCHTAY4FjaLAOyQ6QBHwmYRgB0yIeBQ2CwCsEOmBGyEzSIAO2RSwIGwWQRgh0wL+MvwngB2yMSAQ2GAHTI1YHNfADtkcsDBnQHskOkBswY7ZYLAHsAOmSBwrbCsMwA7FhkO+HGNsKwzADsWGRC4RljWGYAdiwwJXC0s6wzAjkUGBa4UlnUGYMciwwJXCcs6A7BjkYGBK4RlnQHYscjQwOXCss4A7FhkcOBSYVlnAHYsMjxwmbCsMwA7FtkCcImwrDMAOxbZBnDpnAKcZOLAhauwrDMAOxbZCnChsKwzADsW2Q5wkbCsMwA7FtkScIGwrDMAOxbZFvBDYVlnAHYssjXgx573YE4BTqIB2BB72TkFOIkO4MeZ1VjWGYAdi2wZ2FqNZZ0B2LHI1oFTY1lnAHYsMgbg0BhgO+qAw9UY4DQKgYM57eKTAYRjAngbwH7Ju00D3KLIyIArPxgA4AZFRgdcZSwoIxwTwNsE/rJsUy0oIxwTwFsGLl6NBWWEYwJ468BfFqzGgjLCMQE8BuAHxIIywjEBPA5g908GEI4J4LEAZ1ZiQRnhmAAeDbDjG8cLxwTwiICd3ldcOCaAxwScbqYFZYRjAnhUwA5vOy0cE8AjA45WYkEZ4ZgAHhuw9F2JhWMCeHzAwUosKCMcE8AjBBa9aa1wTACPEth0GWC3ItMC9vMvhwC4rsjEgP26t7wUjkkB8Gpm8oV/82I2e3rlL2azY9+/+fRNrsjkgB+8pmlXgU1WH7zxV8/Md3evzlYB8/lxvsj0gGvfL293gO9eBp7nX5hvg1XX/Pv8Kl9kisBVK7FwTDqAF8E6e//Vmfk2WoMzK/CEgavfTm1ngO9Pj43sZ7MngbHZB1sr8LtBpgxcuhILx6QCODyiuvnkzL/7TXhoFYCvwkMtu8hUgSvebWtngBfP4m+C7bP5snoWfGMdR08cuPTNmHYGODq88hPgYG8c6MbYaZEJAxdupoVj0gAc7oLDM6Vo3xucL+lagx+XvFfPrgDH62pwdPXBm+RHTfvgSDhPLByTBmBZkYkDP1iJhWMCeDLAOakl4N8AABBvSURBVGHhmAAGGOCRAGeFhWMCeELAj3NvpwawXQRggEcPbAsLxwTwpIAfZ94QEWC7CMAATwB4IywcE8ATA06FhWMCeGrAjzfvWQuwXQRggKcBHAsLxwTw9IAjYeGYAAYY4LEBP/YAflhEE7ARFo4JYIABHh/wYw/gfBFdwI+lYwIY4MpZXp94QR5dduFQatNJEWXAoXDnwOY11sfmmebRM5B9/8I7jL487wKizKaTItqAzai6BjYvw109OfPPn5l/QZZ7r6Nb0m+6DsBDApuXzt+fHpuXDoQvDFmf7CdzOE+/6zYAl1cRfb6WE7BJsBYb3PD1IbdHh8mvL976tguKYptOiugD9ms+XUsObF5EnUzW4umVBZzuepcADw9cKyzrTGaWF+Y1fAnw+mSzBvd0IA3wsMAL8yYJKfBmw2ztjbsNwFUyX9Z/BKIb8CJ9fWb06tv1SSScfO0+AFcC1wjLOrOZZfM2GCbpaVKwDofnR/PernQAPCTwefiecsfBkXRyoaP3AFwNXC0s6wzXoscMXCks60zVLF8feJ7X11Ws6rt2LQJwA+BleCl62duVaIDrgauEZZ0pn+XkPLivk+CKu3Yvoha4QljWmfJZTq5k9XUZq+Ku3YsA7A7MGjwK4HJhWWdq98H9PRoMsAi4VFjWGY6idxi49wAsAS4TlnUG4PEDlwjLOlN1FO1F4Sh628DFwrLO1M4yFzqUA/f4YBLAUuBCYVln6meZCx0jAC4SlnWmfpb7esad4K7lRQBuDHx71NPzderv2qGIeuACYVln6o+i+9sFA+wA/FBY1hnOg3ce+Ppn7IPHAPxAWNaZilm+8NIcli3TgU0nRXYBOC8s60zVPjhiNWvwf3UhUWLTSRGAmwBHl7DYRI8FOCcs6wwHWTsLzIMNowPOCss6I7jQUftgQ0VxgU3bANwOWHAtGuChgDPCss4IrkXXXsoCeNLA9QF4MGBbWNYZ0UFWzQNKHQEHd/jocu58TQXghsDydAO83Hsd7A2sdwYRZreALWFZZ0YDbJ5nb3b3zg8/7xjwRljWmWGAS95py75rczpmgJ2fQAJwQ+B0H1x7raMWeH3y14XPGihag52fArZrwKmwrDMVa/D1gbnEIXjpSi3w9c/+vfCCdsE+2P2VMgA3BE5efFa/StUCX+z7hUfHD46im7xSZueAE2FZZ7p4+Wgd8O2Hr/1l0Z8J58HbBE7eHav9lSyDW3hJO78PbpTdA46FZZ2p3AcL38KhBjh612mv4DArfxTdKAA3BY5kBGctNcDRMwau/+Lh3jV7kNXwYckdBI6EZZ0Z4Dw4fjPigsOs7Brc8OFngLcMbA6xTArWUA6yGgKHwrLOVMxyvO+sX6Wqgets2mYngY2wrDMVszyXbiy7AU620WyiBwKWP67TDfD80eXFfnz5zCW7CRwIyzpTf6GjPp0Am78nc8Ls/K5NADcE7u2jODbJnwebEyrnp2HvKPCXnqwzFcCNLz2Ik7+SZQ64AR4KWH5iWj5hEpsk5pH++SGbaCmwEW4FLE83wGaXEPxROV/PArglcP02syPghtlZ4GDk7YDFLx8FeJLA8pePdgMsvnBWVGQngX2vJbD05aPdAIsvnBUV2U3gWuGHs9wonQCXXTgzH3obfmxm+Fkwi/ATcKPPdbKLAOwOLH9lQy2weV/ioqsmkgf8V9FnOEWf5nT36sx8SOr5cb7IjgI/rhF+MMvNUgdsHigs/Hi87IWO4gtn51+Y/8efxxb/+/wqXwTgLQJH62btMzqKH2a4/yr8PLb4ExWjNXizAptPTt1p4Brhh7PcKDXA5c/FSe+6/En2d68+mz05Sz8y0+yDsyvwbq/BbYA7u1RZ+IzZyru2Yj5R8e43bzafiWo+pXwVHmrZRXYWuFpYMsuCZ1X2CWwSyFrAq2fxz3YRgBsDC17ZUAdcv4k2MedJF4VvjRntfOOTo2CnHHwTY6dFdhe4UvjhLBek/umsNcDRQVbRaVD+GR3XB/v5R6FX5kPJg51u+qG3q2c+a7ANXCUsAq5/wW4NcPgxxOuTglUzfx5sdgf5v6fgsMpc4Ug+9DZcddkHdwgseL/oOuAATnahw1yt5AXgrsAVwg9m2Y78/aJrgSttksz3b4/MWzi4PlEI4IbA8nQDHPxB7b0uu55VV2SngcuFH85yo3QD3DQANwI2m8pwIy3YJQK8ZeBS4YpZDk5aokscgvdUAHh6wLdHh3x+8HSAy4TLZzk4aZG/hUP5MCQ2bQNwE+Bg9ZW/CUv5MCQ2bQNwuXDFLF+89a14H1w+DImNFfPMD+cdAsDlwoWzHGcpfBe00p5IbTYJXype/thTVRHBZADcOOXDkNhscv1/Atvb/8ulyibAhcJFs9wg5cOQ2JjEDzixBmsFDg7lzK7e7IN5bVJD4CLh/Cw3TPkw4oSXxAqe9px9sKHh56sBvHXg8PmSRU9sz931ReFjinUBuFy4aJYbpHwYYeIT6oKXwGzu2qzje6+DQ3f3y2YANwSWv190+TDClL+4Kb3r9cnzeLnrAx7wbwr8UDg7y4UJnyJVc7WyfBhhBM+qDM+N1r91fivhtIhgMgAuTFfA86K3gi5Yg90DcHWZ/oFjuOQtDfM2YcyLg93fC3xTRDAZABclegpNO+D4IKsSuEUAbgMcnbm0A47eeHp9UrGJbhOAWwDPvffMDrTmqazlw4gTvjtDzctHGwfgxsDXB95huPLWPCRcPgyJTdsA3BT4wvOehzvhujcILR+GxKZtAG4IbJ6G7kfXO2ouEpcPQ2LTNgA3Bf5/0ikuH4bEpm0AbggsT/kwJDZtA3BjYPPwrPvjs642nRQBuAHw0jPPrBS8BfufCyO1aRuAGwKb4+fkI0GrA/Akgc3zpAxw/RPfAZ4o8CFrsGZgc4GSfbBmYP9if30iOYoGeKLA0gAMcIVN2wDcFDj6hAzBpydVA4eXS4qfrwFwS2DJmCrOg4Mj6H3JJ9zVrsFlz7YCeJvAyadUdXCaBPBYgc1/HVzoAHiMwGYTbZ7OWv/WcwBPEthfBsdF80PBW5MBPElg+Qc2AzxJYHkABhjgMQKHT2fu4lo0wKMEDl+QIPnMhlrgSpu2AbghcHL4zOPBSoHlb2UI8CSB5W9lCPAkgeNXrAjeyrCdTSdFAHYHlr9HRzubTooA3GAN7j8ADwwcftxU/FHMQwTgYYHvXhrZ9DPGos8i6/O1KwAPCryY/Tr+CMjog+Pm5mmzb30reM5OK5tOigBcDmw+ZDn85k9XBjb9nE/zXB3z2Sv158HtbDopArBgDY72wRZw8YcJdhiAtwlsnk85b/RZZG42nRQBuAGweTDp0P3DIl1tOikCsBh4c5DVfwDeArB1mtR7AN4GcPJRzAME4IGBhw7AAAuLAAwwwMMHYICFRQAGGODhAzDAwiIAAwzw8AEYYGERgAEGePgADLCwCMAAAzx8AAZYWARggAEePgADLCwCMMAADx/JXd+8mM2Ooy9Pr/xF9IP9vG2AJw189/LYXz0581fhU7XvXp2tAubz43wRgKcKbDzvT4/98y/MT8Gqa/59fpUvAvBUgU2Ctfj+q7Pwu3AN3qzA5qWvAE8eePH06u7VZ7NgQx3ug7MrMGvw5IEXH7zxbz458+9+Ex5aBRvsVXioZRcBeLrAiydn0TfRi1qDw63gG+s4GuBpAy/SV8JFwMHeONCNsdMiAE8V2Gybg6zMZjrc9wbnS6zBioDPZybH5ugqXJXDVZd9sB5gYRGAAQZ4+AAMsLAIwAADPHwABlhYBGCAAR4+AAMsLAIwwAAPH4ABFhYBGGCAhw/AAAuLAAwwwMMHYICFRQAGGODhAzDAwiIAAwzw8AEYYGERgAEGePgADLCwCMAAAzx8AAZYWGRHgSX3BDDAvQVggIVFBCMFePgADLCwiGCkAA8fgAEWFhGMFODhAzDAwiKCkQI8fAAGWFhEMFKAhw/AAAuLCEYK8PABGGBhEcFIAR4+AAMsLCIYKcDDB2CAhUUEIwV4+AAMsLCIYKQADx+AARYWEYwU4OEDMMDCIoKRAjx8AAZYWEQwUoCHD8AAC4sIRqoR+M816W6WmwVggIVFAAYY4OEDMMDCIgADDPDwARhgYRGAAQZ4+AAMsLAIwAADPHwABlhYZDhgwT0JFgHYsYhgMgAePoMCS6a9/SIA2wEYYGGRjqa9/SIA2wEYYGGRjqa9/SIA2wEYYGGRjqa9/SIA2wEYYGGRjqa9/SIA2wEY4CT3p7MP3vj+YjY79v2bT9/kinQ07e0XGTNwPIcDxgH4/Jn5d/fqbPX0yj8/zhfpaNrbLzJm4GgOh4wcOJA162387/OrfJGOpr39IiMGjuewlZhj5MCmY0EHozU4swIDLAWO57ANmGucgcN9sLUCvxvEK0rxb10X6ajMlu/JTJI9hwPGHdg3hwrHq/BQqyrv1pcULNJRmZHc01SAV8+Cb+r2JdOZ9uHuadTAmwOE+6+Mbl1HpzPtw93TqA+yNof4q/BkaeCO6siYT5P8u5fRSXq46tbvg0lBkjkcLlu8iEaGCMDKA7DyjBQ42MU/CQ/Sw0O5mxez2dOrujbbTNxf09HjbTykUJ5xAt98cuavzBzdvTT/Xw174OmeuL93L4/9VQA9/LFyecYJbBJdFv21WYPPv9h2ZwSJH2a7Pz3ewtluecYLvAhm609XZqLuvxr02k/DLKKdSLAWb+F6VXnGCnzzYrMPvnv1WbxHHm+S/hpngEUJ98EhsNnD3f1mJJu80kT9XQT/B1iUaIqSfdloJqw0YQ8X4bE0wJJMEXiRXsvlIKsyZnMXPanATNTmp7Em7qHZl5hwmlSbxSy+VBCuCelPo03Uw/OZyfEWHlIoz0iBSVcBWHkAVh6Alacn4OAwYzarvoJsHnmpO9QUlAmv7rev0k0Z0T0Nm36Ab14cm9FWDdWcWdyfVgsLygQ1/q7mz0RSpZsyonsaOL0A35+GY6w62TePrEUPs7UqY279j+prCqIqnZSR3dPA6QVYMMSV4ERRNFOLZzWPJsrmu4sy45KN0wvwqv7pF4JFRMuYy4LVy0mqdFNGdE9Dpz/g81nVY3xi4Ooy4TLR5r5NlW7KiO5p6PS5ia56gMBhE11V5v40vDpYdXwkqNJRGck9DZ4+D7KqRhqtLtUrjaBMNKeVx2qCKh2VkdzT4OnrNOkLM96qbZV55PT+tHpDXV8mftCm8vhI0JmOyggWGTw9XegIt3nVJ5ar2iXqyyQrS/UGv7YzHZURLTJ0uFSpPAArD8DKA7DyAKw8ACsPwMoDsPJMF3gevcfYfsFNt0fPfX/5PPq645kw8KNL31iGXx4G3ChTB/aXe68LbwY4yuSBrw/M5jjYVh+GPwTfhJtm892j/w6R4xtvj/72yNy4Y5k8sFmDlyHqfoL93Ky9yX/pjbdHwZIXJeu73kwd2NCtT8zaG0gv3/o2vMkGTm+8PTqM1/edyoSBo6PoVC34/+1RJGwDWzc+38U984SB08PnyDBaXZN9cA44+A7giSUPnGx952ZrzBqcRANwupsNf4pX1oJ9MMCTygY4PVAOjZfJGnyYP4oGeFKxgNPzYPM1QA4Z5w/OgwEm+gKw8gCsPAArD8DKA7DyAKw8ACsPwMoDsPL8f2sQ+kfbzVg9AAAAAElFTkSuQmCC" width="100%" />

<p>Please use help(package=“ogrdbstats”) or ? to find function-level
documentation within R.</p>
<h3 id="acknowledgements">Acknowledgements</h3>
<p>Some functions are adapted from <a href="https://tigger.readthedocs.io">TIgGER</a> with thanks to the
authors.</p>
<p>The example annotated reads and inferences linked from this
description are taken from the data of <a href="https://pubmed.ncbi.nlm.nih.gov/?term=27005435">Rubelt et al</a>
and were downloaded from <a href="https://vdjserver.org/">VDJServer</a>.
The genotype was inferred by <a href="https://tigger.readthedocs.io">TIgGER</a>. A small number of
light-chain records were removed from the data set.</p>

</body>
</html>
