<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Overview &mdash; OGRDBstats 0.5.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="install.html" />
    <link rel="prev" title="OGRDBstats" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> OGRDBstats
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#scope-and-features">Scope and Features</a></li>
<li class="toctree-l2"><a class="reference internal" href="#usage-with-genotype-inference-tools">Usage with Genotype Inference tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
<li class="toctree-l2"><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="news.html">Release Notes</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="examples/ungenotyped.html">Using OGRDBstats with any repertoire</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/tigger.html">Using OGRDBstats with TIgGER</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/igdiscover.html">Using OGRDBstats with IgDiscover</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/partis.html">Using OGRDBstats with partis</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/impre.html">Using OGRDBstats with IMPre</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/all_novel.html">Producing a more complete report</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/gapped_v.html">Creating and using a gapped V germline set</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/haplotyping.html">Haplotyping</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tools/ogrdbstats_r.html">ogrdbstats.R - command-line description</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools/formats.html">File Formats</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">OGRDBstats</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Overview</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/overview.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="overview">
<span id="overview-label"></span><h1>Overview<a class="headerlink" href="#overview" title="Permalink to this heading"></a></h1>
<section id="scope-and-features">
<h2>Scope and Features<a class="headerlink" href="#scope-and-features" title="Permalink to this heading"></a></h2>
<p>OGRDBstats can provide allele usage statistics and plots for AIRR-seq repertoires.</p>
<p>The statistics provide overall sequence count for each allele, mutation levels, and recombination statistics, for example the number of J-alleles found
to combine with each allele, and the number of associated CDR3 sequences. An example can be found <a class="reference external" href="https://github.com/airr-community/ogrdbstats/blob/master/example_ogrdbstats_genotype.csv">here</a>.</p>
<p>The available plots provide read pile-up information for each allele, and plots showing the mutation levels and counts associated with each assigned allele. Only
reads with an assignment to a single allele are considered in the analysis. If an allele sequence is equally distant to two or more germline alleles, the aligner
will call these as equally likely in the assignment. Such sequences are not considered in the OGRDBstats analysis, as the analysis is focussed on the level of support
for each allele in the annotated repertoire, and such calls are inconclusive.</p>
<p>The statistics and plots can help to assess the overall level of support for the allele, and the likely valididty of the allele call in the repertoire under study.
For example, if a B-receptor allele is only observed at a low level in the annotated repertoire compared to other alleles of the same gene, and the majority of sequences
assigned to it are mutated compared to the germline, one might suspect that the allele is mis-called, and that the assigned sequences represent mutated sequences
that would be more accurately assigned to another germline allele, that is present in the repertoire at a higher level.</p>
<figure class="align-default" id="id2">
<a class="reference internal image-reference" href="_images/read_pile_up.jpg"><img alt="_images/read_pile_up.jpg" src="_images/read_pile_up.jpg" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-text">Read pile-up for a specific allele. As well as showing the number of supporting alleles and the degree of diversity at specific positions, this can indicate
aretefacts that might otherwise be missed. In this case we can see that some supporting reads are incomplete at the 5-prime end, and some sequences contain
gaps, likely as a result of non-overlapping paired-end reads.</span><a class="headerlink" href="#id2" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-default" id="id3">
<a class="reference internal image-reference" href="_images/mutation_barplot.jpg"><img alt="_images/mutation_barplot.jpg" src="_images/mutation_barplot.jpg" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-text">Mutation bar plot for a specific allele. The format of this plot is modelled after one used by <a class="reference external" href="https://docs.igdiscover.se/en/stable/">IgDiscover</a>.</span><a class="headerlink" href="#id3" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</section>
<section id="usage-with-genotype-inference-tools">
<h2>Usage with Genotype Inference tools<a class="headerlink" href="#usage-with-genotype-inference-tools" title="Permalink to this heading"></a></h2>
<p>A number of tools have been created to infer a subject’s genotype from a repertoire. Because aligners will typically assign each read to the most similar
germline allele, mis-assignments are common, particularly in highly mutated repertoires. These inference tools use a variety of methods to identify
and correct such mis-assignments. This can result in a substantially lower number of alleles being called in the assignments. The set of remaining
alleles represents the subject’s inferred genotype. The tools can also infer ‘novel’ alleles: that is, germline alleles that are not listed in the
germline set, but can be inferred from the repertoire.</p>
<p>The information that OGRDBstats provides can be particularly useful to assess the degree of support for alleles listed in the inferred genotype -
both those listed in the germline set, and those that are inferred. Because the tools provide information in a variety of formats, we have included
specific usage information for tested tools. The tools we have currently tested are:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://docs.igdiscover.se/en/stable/">IgDiscover</a></p></li>
<li><p><a class="reference external" href="https://tigger.readthedocs.io/en/stable/">TIgGER</a></p></li>
<li><p><a class="reference external" href="https://github.com/psathyrella/partis">partis</a></p></li>
<li><p><a class="reference external" href="https://github.com/zhangwei2015/IMPre">IMPre</a></p></li>
</ul>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this heading"></a></h2>
<p>Please cite this paper if you use OGRDBstats in your work:</p>
<p>Lees et al. 2019. OGRDB: A Reference Database of Inferred Immune Receptor Genes. <em>Nucleic Acids Research</em>. <a class="reference external" href="https://doi.org/10.1093/nar/gkz822">doi: 10.1093/nar/gkz822</a>.</p>
<p>Guidance on the use of OGRDBstats statistics and plots to assess the confidence of allele calls can be found in this publication:</p>
<p>Ohlin et al. 2019. Inferred Allelic Variants of Immunoglobulin Receptor Genes: A System for Their Evaluation, Documentation, and Naming. <em>Frontiers in Immunology</em> <a class="reference external" href="https://doi.org/10.3389/fimmu.2019.00435">doi: 10.3389/fimmu.2019.00435</a>.</p>
</section>
<section id="acknowledgements">
<h2>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Permalink to this heading"></a></h2>
<p>Example data is taken from:</p>
<p>Rubelt et al. 2016. Individual Heritable Differences Result in Unique Cell Lymphocyte Receptor Repertoires of Naïve and Antigen-Experienced Cells. <em>Nature Communications</em>. <a class="reference external" href="https://doi.org/10.1038/ncomms11112">doi: 10.1038/ncomms11112</a>.</p>
<p>Some functions in OGRDBstats are adapted from <a class="reference external" href="https://tigger.readthedocs.io/en/stable/">TIgGER</a>, with thanks to the authors.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="OGRDBstats" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="install.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>