Search.setIndex({"docnames": ["examples/all_novel", "examples/gapped_v", "examples/haplotyping", "examples/igdiscover", "examples/impre", "examples/partis", "examples/tigger", "examples/ungenotyped", "index", "install", "news", "overview", "tools/formats", "tools/ogrdbstats_r"], "filenames": ["examples\\all_novel.rst", "examples\\gapped_v.rst", "examples\\haplotyping.rst", "examples\\igdiscover.rst", "examples\\impre.rst", "examples\\partis.rst", "examples\\tigger.rst", "examples\\ungenotyped.rst", "index.rst", "install.rst", "news.rst", "overview.rst", "tools\\formats.rst", "tools\\ogrdbstats_r.rst"], "titles": ["Producing a more complete report", "Creating and using a gapped V germline set", "Haplotyping", "Using OGRDBstats with IgDiscover", "Using OGRDBstats with IMPre", "Using OGRDBstats with partis", "Using OGRDBstats with TIgGER", "Using OGRDBstats with any repertoire", "OGRDBstats", "Installation", "Release Notes", "Overview", "File Formats", "ogrdbstats.R - command-line description"], "terms": {"By": 0, "default": [0, 5, 13], "first": [0, 10], "section": [0, 2, 7], "ogrdbstat": [0, 1, 2, 9, 10, 11, 12], "pdf": [0, 3, 7, 13], "column": [0, 5, 6, 10], "c": 0, "i": [0, 1, 2, 3, 4, 5, 7, 9, 10, 11, 12, 13], "csv": [0, 2, 3, 6, 7], "contain": [0, 1, 3, 5, 6, 7, 11, 12, 13], "data": [0, 5, 6, 11], "novel": [0, 4, 6, 7, 11, 12, 13], "allel": [0, 2, 3, 4, 6, 7, 8, 11, 12, 13], "onli": [0, 1, 5, 6, 11, 12, 13], "These": [0, 1, 6, 11], "can": [0, 1, 2, 3, 4, 5, 6, 7, 9, 11, 12, 13], "extend": 0, "cover": 0, "all": [0, 1, 2, 3, 7, 12, 13], "us": [0, 8, 9, 10, 11, 12, 13], "all_novel": [0, 13], "flag": [0, 13], "which": [0, 2, 5, 7, 10, 12], "instruct": [0, 9], "treat": [0, 13], "though": [0, 9], "thei": [0, 1, 7, 12], "ar": [0, 2, 3, 4, 7, 9, 11, 12, 13], "exampl": [0, 2, 3, 6, 7, 9, 11], "rscript": [0, 3, 5, 6, 7, 9], "r": [0, 3, 5, 6, 7, 8, 9, 10], "v_germline_gap": [0, 6, 7], "fasta": [0, 1, 3, 5, 6, 7, 12, 13], "human": [0, 1, 6, 7, 13], "rep_genotyp": [0, 6, 7], "tsv": [0, 3, 5, 6, 7, 10, 12], "ighv": [0, 1, 3, 6, 7, 13], "process": [0, 3, 5, 7], "time": [0, 1, 5, 12], "size": 0, "mai": [0, 1, 5, 10, 12], "consider": 0, "The": [0, 1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 12, 13], "warn": [0, 1, 12], "ani": [0, 1, 5, 8, 12, 13], "present": [0, 1, 3, 11, 12], "germlin": [0, 3, 4, 5, 6, 7, 8, 11, 12, 13], "set": [0, 3, 4, 5, 6, 7, 8, 11, 12, 13], "observ": [0, 11], "genotyp": [0, 4, 5, 6, 7, 12, 13], "repertoir": [0, 1, 3, 5, 6, 8, 11, 12], "ignor": [0, 1, 12], "one": [0, 1, 10, 11, 13], "would": [0, 3, 11, 13], "expect": [0, 1], "everi": 0, "singl": [0, 11, 13], "If": [1, 2, 3, 7, 9, 11, 12, 13], "sequenc": [1, 3, 4, 5, 6, 7, 10, 11, 12, 13], "you": [1, 2, 3, 5, 7, 9, 11, 12, 13], "annot": [1, 3, 4, 5, 6, 7, 11, 12, 13], "do": [1, 3, 12], "have": [1, 3, 7, 11, 12], "readi": [1, 3], "access": [1, 3], "an": [1, 2, 3, 5, 7, 8, 9, 11, 13], "imgt": [1, 3, 4, 5, 7, 12, 13], "align": [1, 3, 5, 11, 12, 13], "version": [1, 3, 9, 10, 12], "gap_sequ": 1, "command": [1, 3, 5, 6, 7, 8, 9], "from": [1, 2, 3, 4, 5, 10, 11, 12, 13], "receptor_util": [1, 5], "packag": [1, 5, 6, 7, 9, 10], "instal": [1, 8], "requir": [1, 3, 5, 6, 7, 9, 12], "recent": [1, 9, 10], "python": [1, 5], "templat": [1, 12], "download": [1, 3, 4, 5, 9, 12, 13], "follow": [1, 3, 4, 5, 6, 7, 9, 12], "shell": [1, 5], "latest": [1, 9], "pip": 1, "u": 1, "extract_ref": 1, "l": 1, "igh": 1, "homo": [1, 3], "sapien": [1, 3], "homo_sapiens_ighv_gap": 1, "v_gap": 1, "determin": [1, 3, 6, 12], "result": [1, 3, 4, 5, 11, 12], "refer": [1, 9, 12], "provid": [1, 4, 5, 6, 7, 9, 11, 12, 13], "your": [1, 3, 9, 11], "appear": 1, "non": [1, 11], "function": [1, 9, 10, 11], "still": 1, "see": [1, 3, 5, 11, 12, 13], "relat": [1, 12], "few": 1, "file": [1, 3, 4, 5, 6, 7, 8, 10, 13], "complet": [1, 3, 5, 7, 8], "correct": [1, 6, 11], "databas": [1, 3, 4, 11], "ones": 1, "about": 1, "": [1, 4, 5, 6, 11, 12], "k": 1, "canon": 1, "three": 1, "insert": 1, "codon": 1, "rel": 1, "those": [1, 11, 12], "defin": [1, 7], "uniqu": [1, 3, 11, 12], "number": [1, 3, 11, 12], "caus": 1, "problem": 1, "check": [1, 6], "built": 1, "other": [1, 5, 11], "tool": [1, 3, 4, 5, 7, 12], "fix_macaque_gap": 1, "thi": [1, 3, 5, 6, 7, 9, 10, 11, 12, 13], "convert": 1, "fix": 1, "rhesus_macaque_ighv_gap": 1, "rhesus_macaque_ighv_gapped_fix": 1, "replac": 1, "igk": 1, "igl": 1, "notifi": 1, "small": 1, "been": [1, 7, 10, 11, 12], "remov": [1, 5, 13], "becaus": [1, 11], "author": [1, 11], "knowledg": 1, "write": [1, 6, 7, 12], "report": [1, 2, 5, 7, 8, 13], "being": [1, 11], "signific": 1, "level": [1, 11], "airr": [1, 4, 5, 7, 9, 10, 11, 12, 13], "seq": [1, 11], "demonstr": [1, 7], "form": 1, "antibodi": 1, "v": [2, 3, 5, 6, 7, 8, 9, 12, 13], "d": [2, 7, 12], "j": [2, 7, 11, 12], "gene": [2, 5, 6, 7, 11, 12, 13], "rearrang": 2, "taken": [2, 11], "same": [2, 5, 9, 11, 12], "chromosom": 2, "sai": 2, "individu": [2, 6, 11], "ha": [2, 3, 7, 10, 12], "two": [2, 3, 6, 7, 11, 12], "differ": [2, 11, 12], "separ": [2, 4, 12, 13], "consid": [2, 11], "each": [2, 11], "seen": 2, "here": [2, 7, 11, 13], "concern": 2, "call": [2, 4, 6, 7, 11, 12], "anchor": 2, "opportun": 2, "when": [2, 3, 7, 9], "analys": [2, 8, 13], "look": 2, "potenti": 2, "identif": 2, "base": [2, 6, 9, 13], "usag": [2, 5, 13], "summaris": 2, "plot": [2, 5, 7, 11, 13], "3": [2, 10, 13], "where": 2, "discov": [2, 3], "correspond": [2, 12], "gener": 2, "4": 2, "also": [2, 3, 9, 10, 11, 12], "request": 2, "ratio": [2, 12], "ad": [2, 4, 10], "specifi": [2, 3, 10, 13], "option": [2, 5, 7, 10, 13], "hap_gen": [2, 13], "ighj6": [2, 12, 13], "run": [3, 4, 5, 6, 7, 9], "found": [3, 11, 12], "final": [3, 7, 12], "directori": [3, 7, 9], "read": [3, 6, 11, 12, 13], "analysi": [3, 6, 7, 11, 13], "gz": 3, "personalis": 3, "It": [3, 6], "must": [3, 4, 5, 7, 12, 13], "decompress": 3, "befor": [3, 4, 5, 6, 10], "gunzip": 3, "infer": [3, 4, 6, 7, 12, 13], "locat": 3, "For": [3, 7, 9, 11, 12], "accord": [3, 12], "initi": 3, "gap": [3, 5, 7, 8, 10, 11, 12, 13], "wa": [3, 12], "current": [3, 7, 9, 10, 11], "name": [3, 6, 7, 10, 11, 12, 13], "imgt_ref_gap": [3, 5], "wget": 3, "o": [3, 5, 7], "http": [3, 9], "www": 3, "org": 3, "db": 3, "imgtgenedb": 3, "referencesequ": 3, "nt": 3, "withgap": 3, "f": [3, 6, 7, 13], "orf": 3, "inframep": 3, "pleas": [3, 5, 7, 9, 11, 12], "describ": [3, 5], "place": [3, 6], "below": 3, "entir": 3, "multipl": 3, "speci": [3, 7, 12, 13], "without": [3, 5], "further": 3, "enabl": 3, "find": [3, 9], "appripri": 3, "record": [3, 12], "word": [3, 13], "togeth": 3, "space": [3, 13], "output": [3, 4, 5, 6, 7, 10, 13], "inf_fil": [3, 5, 6, 12, 13], "homosapien": [3, 5, 13], "filter": [3, 10, 12], "v_ogrdb_report": 3, "v_ogrdb_plot": 3, "simpler": 3, "doe": [3, 4], "need": [3, 4], "again": [3, 5], "tab": [3, 12, 13], "ighj": [3, 13], "order": 4, "inform": [4, 5, 11], "ogrdb": [4, 11], "One": 4, "possibl": 4, "approach": 4, "igblast": [4, 12], "custom": 4, "includ": [4, 6, 7, 10, 11, 12], "detail": 4, "creat": [4, 5, 6, 7, 8, 11, 12, 13], "given": [4, 7, 12], "setup": 4, "note": [4, 7, 8, 9], "makeblastdb": 4, "select": 4, "format": [4, 5, 7, 8, 10, 11, 13], "along": 4, "although": 5, "some": [5, 11], "yaml": 5, "A": [5, 7, 11, 12], "script": [5, 9, 12], "convert_parti": 5, "py": 5, "combin": [5, 11], "presto": 5, "produc": [5, 6, 7, 8], "chang": [5, 7, 12], "h": [5, 13], "partis_yaml": 5, "partis_tsv": 5, "ogrdb_rec": 5, "ogrdb_v": 5, "posit": [5, 11, 13], "argument": [5, 7, 13], "mode": 5, "v_gene": 5, "help": [5, 9, 10, 11, 13], "show": [5, 11, 13], "messag": [5, 13], "exit": [5, 9, 13], "both": [5, 11, 12], "merg": 5, "extra": 5, "cdr3_seq": 5, "invalid": 5, "in_fram": 5, "stop": 5, "infnam": 5, "tw01a": 5, "outfnam": 5, "n": 5, "proc": 5, "5": [5, 10, 11], "addit": 5, "cach": 5, "fname": 5, "imgt_ref_gapped_dedup": 5, "extract": [5, 6], "tw01a_ogrdb": 5, "tw01a_v_ogrdb": 5, "genoty": 5, "vh": 5, "twice": 5, "onc": [5, 9], "paramet": [5, 7], "remain": [5, 11], "so": 5, "overal": [5, 11, 12], "impact": 5, "low": [5, 11], "suppli": [5, 10], "abl": 5, "obtain": 5, "sourc": 5, "otherwis": [5, 9, 11], "ungap": [5, 10, 12], "error": 5, "duplic": 5, "ident": 5, "successfulli": 5, "identical_sequ": 5, "identifi": [5, 6, 11, 12], "To": [6, 9], "conduct": [6, 7], "findnovelallel": 6, "infergenotyp": 6, "infergenotypebayesian": 6, "reassignallel": 6, "code": [6, 7], "vignett": [6, 10], "perform": 6, "step": 6, "save": [6, 7], "sampl": [6, 7], "load": 6, "librari": [6, 7], "dplyr": 6, "writefasta": [6, 7], "samplegermlineighv": [6, 7], "detect": 6, "airrdb": 6, "nproc": 6, "1": [6, 10], "row": [6, 7], "success": 6, "novel_row": 6, "selectnovel": 6, "unmut": [6, 12, 13], "earlier": 6, "geno": 6, "germline_db": 6, "find_unmut": 6, "true": 6, "vector": 6, "genotype_db": 6, "genotypefasta": 6, "personl": 6, "assign": [6, 11, 12, 13], "updat": 6, "v_call_genotyp": [6, 12], "sample_db": 6, "tabl": [6, 7], "sep": [6, 7], "t": [6, 7], "v_genotyped_seq": 6, "origin": 6, "referenc": 6, "rep_genotyped_ogrdb_plot": [6, 7], "rep_genotyped_ogrdb_report": [6, 7], "how": 7, "therefor": 7, "we": [7, 9, 11], "tigger": [7, 8, 11], "match": [7, 12], "docker": [7, 10], "imag": [7, 10], "pwd": [7, 9], "scratch": [7, 9], "williamle": [7, 9], "stabl": [7, 9], "make_sample_data": 7, "With": 7, "should": [7, 12, 13], "run_ogrdbstat": [7, 9, 13], "statu": 7, "statist": [7, 11, 13], "filenam": [7, 13], "case": [7, 10, 11], "7": 7, "loci": [7, 12], "support": [7, 10, 11], "full": [7, 9], "descript": [7, 8, 12], "avail": [7, 9, 10, 11], "line": [7, 8, 9], "recombin": [8, 11], "adapt": [8, 11], "immun": [8, 11], "receptor": [8, 11], "overview": 8, "releas": 8, "igdiscov": [8, 10, 11, 12, 13], "parti": [8, 11], "impr": [8, 11, 13], "more": [8, 10, 11], "haplotyp": [8, 12, 13], "index": 8, "search": 8, "page": 8, "fulli": 9, "configur": 9, "develop": 9, "pandoc": 9, "rstudio": 9, "machin": 9, "alreadi": 9, "websit": 9, "start": 9, "interpret": 9, "altern": 9, "github": 9, "devtool": 9, "install_github": 9, "com": 9, "commun": [9, 11], "choic": 9, "via": 9, "put": 9, "wherev": 9, "want": 9, "its": [9, 10, 12], "path": 9, "In": [9, 11, 12], "document": [9, 10, 11], "alwai": 9, "silent": 9, "cannot": 9, "hub": 9, "pull": 9, "build": 9, "linux": 9, "mac": 9, "Or": 9, "window": 9, "prompt": 9, "cd": 9, "exactli": [9, 12], "wai": 9, "abov": 9, "try": 9, "0": 10, "file_prefix": [10, 13], "allow": 10, "user": 10, "prefix": [10, 13], "2": 10, "now": 10, "resolv": 10, "correctli": 10, "knitr": 10, "even": 10, "input": 10, "than": 10, "period": 10, "cran": 10, "rework": 10, "extern": 10, "callabl": 10, "count": [11, 12], "mutat": 11, "associ": [11, 12], "cdr3": [11, 12], "pile": 11, "up": 11, "equal": [11, 13], "distant": 11, "like": 11, "Such": 11, "focuss": 11, "inconclus": 11, "assess": 11, "valididti": 11, "under": 11, "studi": 11, "b": 11, "compar": 11, "major": 11, "might": 11, "suspect": 11, "mi": 11, "repres": 11, "accur": 11, "anoth": 11, "higher": 11, "specif": [11, 12], "As": [11, 12], "well": 11, "degre": 11, "divers": 11, "indic": 11, "aretefact": 11, "miss": 11, "incomplet": 11, "prime": 11, "end": 11, "overlap": 11, "pair": 11, "bar": 11, "model": 11, "after": [11, 12], "subject": [11, 12], "typic": 11, "most": 11, "similar": 11, "common": 11, "particularli": 11, "highli": 11, "varieti": 11, "method": 11, "substanti": 11, "lower": 11, "list": [11, 12, 13], "test": 11, "cite": 11, "paper": 11, "work": 11, "lee": 11, "et": 11, "al": 11, "2019": 11, "nucleic": 11, "acid": [11, 12], "research": 11, "doi": 11, "10": 11, "1093": 11, "nar": 11, "gkz822": 11, "guidanc": 11, "confid": 11, "public": 11, "ohlin": 11, "variant": 11, "immunoglobulin": 11, "system": 11, "Their": 11, "evalu": 11, "frontier": 11, "immunologi": 11, "3389": 11, "fimmu": 11, "00435": 11, "rubelt": 11, "2016": 11, "herit": 11, "cell": 11, "lymphocyt": 11, "na\u00efv": 11, "antigen": 11, "experienc": 11, "natur": 11, "1038": 11, "ncomms11112": 11, "thank": 11, "ref_fil": [12, 13], "apart": 12, "simpli": 12, "consist": 12, "unless": 12, "next": 12, "point": 12, "out": 12, "nomin": 12, "make": 12, "sure": 12, "read_fil": [12, 13], "changeo": [12, 13], "automat": 12, "least": 12, "sequence_id": 12, "d_call": 12, "j_call": 12, "sequence_align": 12, "j_sequence_start": 12, "j_sequence_end": 12, "j_germline_start": 12, "j_germline_end": 12, "equival": 12, "field": [12, 13], "compat": 12, "sequence_imgt": 12, "cdr3_imgt": 12, "v_mut_nc": 12, "d_mut_nc": 12, "j_mut_nc": 12, "junction_start": 12, "v_seq": 12, "d_seq": 12, "j_seq": 12, "heavi": 12, "chain": [12, 13], "made": 12, "either": 12, "unagp": 12, "fall": 12, "back": 12, "type": [12, 13], "locu": 12, "header": [12, 13], "sole": 12, "nearest": 12, "satisfi": 12, "just": 12, "prefer": 12, "presenc": 12, "easier": 12, "label": 12, "closest_refer": 12, "closest": 12, "closest_host": 12, "nt_diff": 12, "nucleotid": 12, "between": 12, "nt_diff_host": 12, "nt_substitut": 12, "comma": 12, "substitut": 12, "e": [12, 13], "g": [12, 13], "g112a": 12, "aa_diff": 12, "amino": 12, "aa_substitut": 12, "a96n": 12, "unmutated_sequ": 12, "assigned_unmutated_frequ": 12, "divid": 12, "x": [12, 13], "100": 12, "unmutated_umi": 12, "molecul": 12, "molecular": 12, "umi": 12, "were": 12, "allelic_percentag": 12, "unmutated_frequ": 12, "unique_v": 12, "unique_d": 12, "unique_j": 12, "unique_cdr3": 12, "unique_vs_unmut": 12, "unique_ds_unmut": 12, "unique_js_unmut": 12, "unique_cdr3s_unmut": 12, "haplotyping_gen": 12, "haplotyping_ratio": 12, "express": 12, "percentag": 12, "60": 12, "40": 12, "nt_sequenc": 12, "consensu": 12, "appli": 13, "dockeris": 13, "plot_unmut": 13, "opt": 13, "igkv": 13, "iglv": 13, "ighd": 13, "igkj": 13, "iglj": 13, "trav": 13, "traj": 13, "trbv": 13, "trbd": 13, "trbj": 13, "trgv": 13, "trgj": 13, "trdv": 13, "trdd": 13, "trdj": 13, "p": 13, "composit": 13, "rd": 13, "valu": 13, "html": 13, "none": 13, "pathnam": 13, "usual": 13}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"produc": 0, "more": 0, "complet": 0, "report": [0, 3, 12], "creat": [1, 3], "us": [1, 3, 4, 5, 6, 7], "gap": 1, "v": 1, "germlin": 1, "set": 1, "rhesu": 1, "macaqu": 1, "haplotyp": 2, "ogrdbstat": [3, 4, 5, 6, 7, 8, 13], "igdiscov": 3, "d": 3, "j": 3, "gene": 3, "impr": 4, "parti": 5, "tigger": 6, "ani": 7, "repertoir": 7, "get": 8, "start": 8, "exampl": 8, "usag": [8, 11], "document": 8, "indic": 8, "tabl": 8, "instal": 9, "from": 9, "cran": 9, "docker": 9, "imag": 9, "releas": 10, "note": 10, "overview": 11, "scope": 11, "featur": 11, "genotyp": 11, "infer": 11, "tool": 11, "refer": 11, "acknowledg": 11, "file": 12, "format": 12, "input": 12, "detail": 12, "column": 12, "statist": 12, "r": 13, "command": 13, "line": 13, "descript": 13}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"Producing a more complete report": [[0, "producing-a-more-complete-report"]], "Creating and using a gapped V germline set": [[1, "creating-and-using-a-gapped-v-germline-set"]], "Creating and using a gapped V germline set for Rhesus macaque": [[1, "creating-and-using-a-gapped-v-germline-set-for-rhesus-macaque"]], "Haplotyping": [[2, "haplotyping"]], "Using OGRDBstats with IgDiscover": [[3, "using-ogrdbstats-with-igdiscover"]], "Creating reports for D- or J-genes": [[3, "creating-reports-for-d-or-j-genes"]], "Using OGRDBstats with IMPre": [[4, "using-ogrdbstats-with-impre"]], "Using OGRDBstats with partis": [[5, "using-ogrdbstats-with-partis"]], "Using OGRDBstats with TIgGER": [[6, "using-ogrdbstats-with-tigger"]], "Using OGRDBstats with any repertoire": [[7, "using-ogrdbstats-with-any-repertoire"]], "OGRDBstats": [[8, "ogrdbstats"]], "Getting Started": [[8, null]], "Examples": [[8, null]], "Usage Documentation": [[8, null]], "Indices and tables": [[8, "indices-and-tables"]], "Installation": [[9, "installation"]], "Installing from CRAN": [[9, "installing-from-cran"]], "Docker Image": [[9, "id1"]], "Release Notes": [[10, "release-notes"]], "Overview": [[11, "overview"]], "Scope and Features": [[11, "scope-and-features"]], "Usage with Genotype Inference tools": [[11, "usage-with-genotype-inference-tools"]], "References": [[11, "references"]], "Acknowledgements": [[11, "acknowledgements"]], "File Formats": [[12, "file-formats"]], "Input file formats in detail": [[12, "input-file-formats-in-detail"]], "Columns in the statistics report": [[12, "columns-in-the-statistics-report"]], "ogrdbstats.R - command-line description": [[13, "ogrdbstats-r-command-line-description"]]}, "indexentries": {}})