Search.setIndex({"docnames": ["examples/all_novel", "examples/gapped_v", "examples/igdiscover", "examples/impre", "examples/partis", "examples/tigger", "examples/ungenotyped", "index", "install", "news", "overview", "tools/formats", "tools/ogrdbstats_r"], "filenames": ["examples\\all_novel.rst", "examples\\gapped_v.rst", "examples\\igdiscover.rst", "examples\\impre.rst", "examples\\partis.rst", "examples\\tigger.rst", "examples\\ungenotyped.rst", "index.rst", "install.rst", "news.rst", "overview.rst", "tools\\formats.rst", "tools\\ogrdbstats_r.rst"], "titles": ["Producing a more complete report", "Creating and using a gapped V germline set", "Using OGRDBstats with IgDiscover", "Using OGRDBstats with IMPre", "Using OGRDBstats with partis", "Using OGRDBstats with TIgGER", "Using OGRDBstats with any repertoire", "OGRDBstats: analyse gene and allele usage in adaptive immune receptor repertoires", "Installation", "Release Notes", "Overview", "Detailed Description of Input Files", "ogrdbstats.R - command-line description"], "terms": {"The": [0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12], "first": [0, 9], "section": [0, 6, 12], "ogrdbstat": [0, 1, 8, 9, 10, 11], "pdf": [0, 2, 6, 12], "column": [0, 4, 5, 11], "c": 0, "i": [0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12], "csv": [0, 2, 5, 6], "default": [0, 4, 12], "contain": [0, 1, 2, 4, 5, 6, 10, 11, 12], "data": [0, 4, 5, 10], "novel": [0, 3, 5, 6, 10, 11, 12], "allel": [0, 2, 3, 5, 6, 10, 11, 12], "onli": [0, 1, 4, 5, 10, 11, 12], "These": [0, 1, 5, 10], "can": [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12], "extend": 0, "cover": 0, "all": [0, 1, 2, 6, 11, 12], "us": [0, 7, 8, 10, 11, 12], "all_novel": [0, 12], "flag": [0, 12], "which": [0, 1, 4, 6], "instruct": [0, 8], "treat": [0, 12], "though": [0, 8], "thei": [0, 1, 6, 11], "ar": [0, 2, 3, 6, 8, 10, 11, 12], "exampl": [0, 2, 5, 6, 8, 10], "rscript": [0, 2, 4, 5, 6, 8], "r": [0, 2, 4, 5, 6, 7, 8, 9], "v_germline_gap": [0, 5, 6], "fasta": [0, 1, 2, 4, 5, 6, 11, 12], "human": [0, 1, 5, 6, 12], "rep_genotyp": [0, 5, 6], "tsv": [0, 2, 4, 5, 6, 11], "ighv": [0, 1, 2, 5, 6, 12], "process": [0, 2, 4, 6], "time": [0, 1, 4, 11], "size": 0, "consider": 0, "warn": [0, 1, 11], "ani": [0, 1, 4, 7, 10, 11, 12], "present": [0, 1, 2, 10, 11], "germlin": [0, 2, 3, 4, 5, 6, 7, 10, 11, 12], "set": [0, 2, 3, 4, 5, 6, 7, 10, 11, 12], "observ": [0, 10], "genotyp": [0, 3, 4, 5, 6, 11, 12], "repertoir": [0, 1, 2, 4, 5, 10, 11], "ignor": [0, 1, 11], "one": [0, 1, 10, 12], "would": [0, 2, 10, 12], "expect": [0, 1], "everi": 0, "singl": [0, 10, 12], "mai": [0, 1, 4, 10, 11], "silenc": [], "later": [], "version": [1, 2, 8, 9, 11], "If": [1, 2, 6, 8, 10, 11, 12], "sequenc": [1, 2, 3, 4, 5, 6, 7, 10, 11, 12], "you": [1, 2, 4, 6, 8, 10, 11, 12], "annot": [1, 2, 3, 4, 5, 6, 10, 11, 12], "do": [1, 2, 11], "have": [1, 2, 6, 10, 11], "readi": [1, 2], "access": [1, 2], "an": [1, 2, 4, 6, 7, 8, 10, 12], "imgt": [1, 2, 3, 4, 6, 11, 12], "align": [1, 2, 4, 10, 11, 12], "gap_sequ": 1, "command": [1, 2, 4, 5, 6, 7, 8], "from": [1, 2, 3, 4, 9, 10, 11, 12], "receptor_util": [1, 4], "packag": [1, 4, 5, 6, 7, 8, 9], "instal": [1, 7], "requir": [1, 2, 4, 5, 6, 8, 11], "recent": [1, 8], "python": [1, 4], "templat": [1, 11], "download": [1, 2, 3, 4, 8, 11, 12], "follow": [1, 2, 3, 4, 5, 6, 8, 11], "latest": [1, 6, 8], "pip": 1, "u": 1, "extract_ref": 1, "l": 1, "igh": 1, "homo": [1, 2], "sapien": [1, 2], "homo_sapiens_ighv_gap": 1, "v_gap": 1, "determin": [1, 2, 5, 11], "result": [1, 2, 3, 4, 10, 11], "provid": [1, 3, 4, 5, 6, 8, 10, 11, 12], "your": [1, 2, 8, 10], "appear": 1, "non": [1, 10], "function": [1, 8, 9, 10], "still": 1, "see": [1, 2, 4, 10, 11, 12], "relat": [1, 11], "few": 1, "complet": [1, 2, 4, 6, 7], "correct": [1, 5, 10], "databas": [1, 2, 3, 10], "ones": 1, "about": 1, "": [1, 3, 4, 5, 10, 11], "k": 1, "canon": 1, "three": 1, "insert": 1, "codon": 1, "rel": 1, "those": [1, 10, 11], "defin": [1, 6], "uniqu": [1, 2, 10, 11], "number": [1, 2, 10, 11], "caus": 1, "problem": 1, "check": [1, 5], "built": 1, "other": [1, 4, 10], "tool": [1, 2, 3, 4, 6, 11], "fix_macaque_gap": 1, "thi": [1, 2, 4, 5, 6, 8, 9, 10, 11, 12], "fix": 1, "rhesus_macaque_ighv_gap": 1, "rhesus_macaque_ighv_gapped_fix": 1, "replac": 1, "igk": 1, "igl": 1, "notifi": 1, "small": 1, "been": [1, 6, 9, 10, 11], "remov": [1, 4, 12], "becaus": [1, 10], "author": [1, 10], "knowledg": 1, "write": [1, 5, 6, 11], "report": [1, 4, 6, 7, 12], "being": [1, 10], "signific": 1, "level": [1, 10], "airr": [1, 3, 4, 6, 8, 10, 11, 12], "seq": 1, "demonstr": [1, 6], "form": 1, "antibodi": 1, "when": [2, 6, 8], "run": [2, 3, 4, 6, 8], "ha": [2, 6, 9, 11], "found": [2, 10], "final": [2, 6, 11], "directori": [2, 6, 8], "file": [1, 2, 3, 4, 5, 6, 7, 12], "analysi": [2, 5, 6, 7, 10, 12], "gz": 2, "read": [2, 5, 10, 11, 12], "personalis": 2, "It": [2, 5], "must": [2, 3, 4, 6, 11, 12], "decompress": 2, "befor": [2, 3, 4], "unzip": 2, "also": [2, 8, 9, 10, 11], "infer": [2, 3, 5, 6, 11, 12], "locat": 2, "v": [2, 4, 5, 6, 7, 8, 11, 12], "For": [2, 6, 8, 10, 11], "refer": [1, 8], "accord": [2, 11], "initi": 2, "gap": [2, 4, 6, 7, 10, 11, 12], "wa": 2, "otherwis": [4, 8, 10], "pleas": [2, 4, 6, 8, 10, 11], "describ": [2, 4], "specifi": [2, 12], "place": [2, 5], "imgt_ref_gap": [2, 4], "name": [2, 5, 6, 10, 11, 12], "wget": 2, "o": [2, 4, 5, 6, 10], "http": [2, 8], "www": 2, "org": 2, "db": 2, "imgtgenedb": 2, "referencesequ": 2, "nt": 2, "withgap": 2, "f": [2, 5, 6, 12], "orf": 2, "inframep": 2, "entir": 2, "multipl": 2, "speci": [2, 6, 11, 12], "without": [2, 4], "further": 2, "enabl": 2, "find": [2, 8], "appripri": 2, "record": [2, 11], "two": [2, 5, 6, 10], "word": [2, 12], "togeth": 2, "space": [2, 12], "output": [2, 3, 4, 5, 6, 12], "inf_fil": [2, 4, 5, 11, 12], "homosapien": [2, 4, 12], "filter": [2, 11], "v_ogrdb_report": 2, "v_ogrdb_plot": 2, "discov": 2, "simpler": 2, "doe": [2, 3], "need": [2, 3], "again": [2, 4], "tab": [2, 11, 12], "ighj": [2, 12], "call": [3, 5, 6, 10, 11], "separ": [3, 11, 12], "order": 3, "inform": [3, 4, 10], "ogrdb": [3, 10], "One": 3, "possibl": 3, "approach": 3, "igblast": [3, 11], "custom": 3, "includ": [3, 5, 6, 10, 11], "detail": [3, 7], "creat": [3, 4, 5, 6, 7, 10, 11, 12], "given": [3, 6, 11], "setup": 3, "note": [3, 6, 7, 8], "ad": [3, 9], "makeblastdb": 3, "select": 3, "format": [3, 4, 5, 6, 10, 11, 12], "along": 3, "although": 4, "some": [4, 10], "yaml": 4, "A": [4, 6, 10, 11], "script": [4, 8, 11], "convert_parti": 4, "py": 4, "combin": [4, 10], "presto": 4, "produc": [4, 5, 6, 7], "chang": [4, 5, 6, 10, 11], "usag": [4, 12], "h": [4, 12], "partis_yaml": 4, "partis_tsv": 4, "ogrdb_rec": 4, "ogrdb_v": 4, "posit": [4, 10, 12], "argument": [4, 6, 12], "mode": 4, "v_gene": 4, "option": [4, 6, 12], "help": [4, 8, 9, 10, 12], "show": [4, 10, 12], "messag": [4, 12], "exit": [4, 8, 12], "shell": [1, 4], "both": [4, 10, 11], "merg": 4, "extra": 4, "cdr3_seq": 4, "invalid": 4, "in_fram": 4, "stop": 4, "infnam": 4, "tw01a": 4, "outfnam": 4, "n": 4, "proc": 4, "5": [4, 9, 10], "addit": 4, "cach": 4, "fname": 4, "imgt_ref_gapped_dedup": 4, "extract": [4, 5], "tw01a_ogrdb": 4, "tw01a_v_ogrdb": 4, "genoty": 4, "plot": [4, 6, 10, 12], "vh": 4, "twice": 4, "onc": [4, 8], "paramet": [4, 6], "remain": [4, 10], "same": [4, 8, 10, 11], "so": 4, "overal": [4, 10], "impact": 4, "low": [4, 10], "suppli": 4, "gene": [4, 5, 6, 10, 11, 12], "abl": 4, "obtain": 4, "sourc": 4, "ungap": [4, 11], "error": 4, "duplic": 4, "ident": 4, "successfulli": 4, "identical_sequ": 4, "identifi": [4, 5, 10], "To": [5, 8], "conduct": [5, 6], "findnovelallel": 5, "infergenotyp": 5, "infergenotypebayesian": 5, "reassignallel": 5, "base": [5, 8, 12], "line": [6, 7, 8], "vignett": 5, "perform": 5, "step": 5, "save": [5, 6], "sampl": [5, 6], "load": 5, "librari": [5, 6], "dplyr": 5, "writefasta": [5, 6], "samplegermlineighv": [5, 6], "detect": 5, "airrdb": 5, "nproc": 5, "1": 5, "row": [5, 6], "success": 5, "novel_row": 5, "selectnovel": 5, "individu": [5, 10], "unmut": [5, 11, 12], "earlier": 5, "geno": 5, "germline_db": 5, "find_unmut": 5, "true": 5, "vector": 5, "genotype_db": 5, "genotypefasta": 5, "personl": 5, "assign": [5, 10, 11, 12], "updat": 5, "v_call_genotyp": [5, 11], "sample_db": 5, "tabl": [5, 6], "sep": [5, 6], "t": [5, 6], "v_genotyped_seq": 5, "code": [5, 6], "referenc": 5, "rep_genotyped_ogrdb_plot": [5, 6], "rep_genotyped_ogrdb_report": [5, 6], "how": 6, "therefor": 6, "alleleshav": [], "we": [6, 8, 10], "tigger": [6, 7, 10], "match": 6, "here": [6, 10, 12], "current": [2, 6, 8, 10], "docker": [6, 9], "imag": [6, 9], "pwd": [6, 8], "scratch": [6, 8], "make_sample_data": 6, "With": 6, "should": [6, 11, 12], "case": [6, 10], "statu": 6, "statist": [6, 10, 12], "filenam": [6, 12], "d": [6, 11], "j": [6, 10, 11], "7": 6, "loci": 6, "support": [6, 10], "full": [6, 8], "descript": [6, 7], "avail": [6, 8, 9, 10], "util": 7, "associ": [7, 10], "overview": 7, "new": [], "igdiscov": [7, 10, 11, 12], "parti": [7, 10], "impr": [7, 10, 12], "more": [7, 10], "input": 7, "index": 7, "search": 7, "page": 7, "fulli": 8, "configur": 8, "stabl": 8, "develop": 8, "pandoc": 8, "rstudio": 8, "machin": 8, "alreadi": 8, "websit": 8, "start": 8, "interpret": 8, "altern": 8, "github": 8, "devtool": 8, "install_github": 8, "com": 8, "commun": [8, 10], "choic": 8, "via": 8, "put": 8, "wherev": 8, "want": 8, "its": [8, 11], "path": 8, "In": [8, 10, 11], "document": [8, 9, 10], "alwai": 8, "silent": 8, "cannot": 8, "hub": 8, "williamle": 8, "pull": 8, "build": 8, "linux": 8, "mac": 8, "run_ogrdbstat": [8, 12], "Or": 8, "window": 8, "prompt": 8, "cd": 8, "exactli": 8, "wai": 8, "abov": 8, "try": 8, "0": 9, "submit": [], "cran": 9, "either": [10, 11], "legaci": 10, "count": [10, 11], "each": 10, "mutat": 10, "recombin": 10, "cdr3": [10, 11], "pile": 10, "up": 10, "consid": 10, "equal": [10, 12], "distant": 10, "stwo": [], "like": 10, "Such": 10, "focuss": 10, "inconclus": 10, "assess": 10, "valididti": 10, "under": 10, "studi": 10, "b": 10, "receptor": 10, "compar": 10, "major": 10, "might": 10, "suspect": 10, "mi": 10, "repres": 10, "accur": 10, "anoth": 10, "higher": 10, "specif": 10, "As": [10, 11], "well": 10, "degre": 10, "divers": 10, "indic": 10, "aretefact": 10, "miss": 10, "incomplet": 10, "prime": 10, "end": 10, "overlap": 10, "pair": 10, "bar": 10, "model": 10, "after": [10, 11], "subject": [10, 11], "typic": 10, "most": 10, "similar": 10, "common": 10, "particularli": 10, "highli": 10, "varieti": 10, "method": 10, "substanti": 10, "lower": 10, "list": [10, 12], "test": 10, "cite": 10, "paper": 10, "work": 10, "lee": 10, "et": 10, "al": 10, "2019": 10, "immun": 10, "nucleic": 10, "acid": 10, "research": 10, "doi": 10, "10": 10, "1093": 10, "nar": 10, "gkz822": 10, "guidanc": 10, "confid": 10, "public": 10, "ohlin": 10, "variant": 10, "immunoglobulin": 10, "system": 10, "Their": 10, "evalu": 10, "frontier": 10, "immunologi": 10, "3389": 10, "fimmu": 10, "00435": 10, "taken": 10, "rubelt": 10, "2016": 10, "herit": 10, "differ": 10, "cell": 10, "lymphocyt": 10, "na\u00efv": 10, "antigen": 10, "experienc": 10, "natur": 10, "1038": 10, "ncomms11112": 10, "adapt": 10, "thank": 10, "ref_fil": [11, 12], "apart": 11, "simpli": 11, "consist": 11, "unless": 11, "next": 11, "point": 11, "out": 11, "nomin": 11, "make": 11, "sure": 11, "correspond": 11, "read_fil": [11, 12], "miairr": [], "changeo": [11, 12], "automat": 11, "least": 11, "sequence_id": 11, "d_call": 11, "j_call": 11, "sequence_align": 11, "j_sequence_start": 11, "j_sequence_end": 11, "j_germline_start": 11, "j_germline_end": 11, "equival": 11, "field": [11, 12], "compat": 11, "sequence_imgt": 11, "cdr3_imgt": 11, "v_mut_nc": 11, "d_mut_nc": 11, "j_mut_nc": 11, "junction_start": 11, "v_seq": 11, "d_seq": 11, "j_seq": 11, "quest": [], "pars": [], "them": [], "convert": [], "heavi": 11, "chain": [11, 12], "made": 11, "unagp": 11, "recommend": [], "fall": 11, "back": 11, "type": [11, 12], "locu": 11, "header": [11, 12], "nearest": 11, "satisfi": 11, "just": 11, "prefer": 11, "presenc": 11, "easier": 11, "appli": 12, "dockeris": 12, "plot_unmut": 12, "opt": 12, "hap_gen": 12, "3": 12, "e": 12, "g": 12, "igkv": 12, "iglv": 12, "ighd": 12, "igkj": 12, "iglj": 12, "trav": 12, "traj": 12, "trbv": 12, "trbd": 12, "trbj": 12, "trgv": 12, "trgj": 12, "trdv": 12, "trdd": 12, "trdj": 12, "p": 12, "composit": 12, "x": 12, "rd": 12, "valu": 12, "haplotyp": 12, "ighj6": 12, "html": 12, "none": 12, "pathnam": 12, "usual": 12, "analys": 12, "below": [2, 12], "releas": 7, "gape": [], "rework": 9, "extern": 9, "callabl": 9, "By": 0, "sole": 11}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"produc": 0, "more": 0, "complet": 0, "report": [0, 2], "creat": [1, 2], "us": [1, 2, 3, 4, 5, 6], "gap": 1, "v": 1, "refer": 10, "file": 11, "rhesu": 1, "macaqu": 1, "ogrdbstat": [2, 3, 4, 5, 6, 7, 12], "igdiscov": 2, "d": 2, "j": 2, "gene": [2, 7], "impr": 3, "parti": 4, "tigger": 5, "ani": 6, "repertoir": [6, 7], "analys": 7, "allel": 7, "usag": [7, 10], "adapt": 7, "immun": 7, "receptor": 7, "get": 7, "start": 7, "exampl": 7, "document": 7, "indic": 7, "tabl": 7, "instal": 8, "from": 8, "cran": 8, "docker": 8, "imag": 8, "new": [], "overview": 10, "scope": 10, "featur": 10, "genotyp": 10, "infer": 10, "tool": 10, "acknowledg": 10, "detail": 11, "descript": [11, 12], "input": 11, "r": 12, "command": 12, "line": 12, "releas": 9, "note": 9, "germlin": 1, "set": 1}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"Using OGRDBstats with IMPre": [[3, "using-ogrdbstats-with-impre"]], "OGRDBstats: analyse gene and allele usage in adaptive immune receptor repertoires": [[7, "ogrdbstats-analyse-gene-and-allele-usage-in-adaptive-immune-receptor-repertoires"]], "Getting Started": [[7, null]], "Examples": [[7, null]], "Usage Documentation": [[7, null]], "Indices and tables": [[7, "indices-and-tables"]], "Installation": [[8, "installation"]], "Installing from CRAN": [[8, "installing-from-cran"]], "Docker Image": [[8, "id1"]], "Release Notes": [[9, "release-notes"]], "Overview": [[10, "overview"]], "Scope and Features": [[10, "scope-and-features"]], "Usage with Genotype Inference tools": [[10, "usage-with-genotype-inference-tools"]], "References": [[10, "references"]], "Acknowledgements": [[10, "acknowledgements"]], "Using OGRDBstats with TIgGER": [[5, "using-ogrdbstats-with-tigger"]], "Using OGRDBstats with any repertoire": [[6, "using-ogrdbstats-with-any-repertoire"]], "Producing a more complete report": [[0, "producing-a-more-complete-report"]], "Creating and using a gapped V germline set": [[1, "creating-and-using-a-gapped-v-germline-set"]], "Creating and using a gapped V germline set for Rhesus macaque": [[1, "creating-and-using-a-gapped-v-germline-set-for-rhesus-macaque"]], "Using OGRDBstats with IgDiscover": [[2, "using-ogrdbstats-with-igdiscover"]], "Creating reports for D- or J-genes": [[2, "creating-reports-for-d-or-j-genes"]], "Using OGRDBstats with partis": [[4, "using-ogrdbstats-with-partis"]], "Detailed Description of Input Files": [[11, "detailed-description-of-input-files"]], "ogrdbstats.R - command-line description": [[12, "ogrdbstats-r-command-line-description"]]}, "indexentries": {}})